Index: ompi/mca/btl/scif/btl_scif_module.c
===================================================================
--- ompi/mca/btl/scif/btl_scif_module.c	(revision 31713)
+++ ompi/mca/btl/scif/btl_scif_module.c	(working copy)
@@ -2,6 +2,8 @@
 /*
  * Copyright (c) 2013      Los Alamos National Security, LLC. All rights
  *                         reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  *
  * Additional copyrights may follow
@@ -314,6 +316,7 @@
                 mca_btl_scif_frag_return (frag);
                 return NULL;
             }
+            *size = max_size;
         }
 
         frag->segments[0].base.seg_len = reserve + *size;

