diff -ENwbur /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8/ompi/errhandler/errcode.c /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8_patched/ompi/errhandler/errcode.c
--- /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8/ompi/errhandler/errcode.c	2012-10-05 15:13:24.000000000 +0400
+++ /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8_patched/ompi/errhandler/errcode.c	2013-03-25 12:40:43.000000000 +0400
@@ -172,8 +172,8 @@
     CONSTRUCT_ERRCODE( ompi_err_unsupported_operation, MPI_ERR_UNSUPPORTED_OPERATION, "MPI_ERR_UNSUPPORTED_OPERATION: operation not supported" );
     CONSTRUCT_ERRCODE( ompi_err_win, MPI_ERR_WIN, "MPI_ERR_WIN: invalid window" );
 
-    ompi_mpi_errcode_lastused = MPI_ERR_WIN;
-    ompi_mpi_errcode_lastpredefined = MPI_ERR_WIN;
+    ompi_mpi_errcode_lastused = MPI_ERR_LASTCODE;
+    ompi_mpi_errcode_lastpredefined = MPI_ERR_LASTCODE;
     return OMPI_SUCCESS;
 }
 
diff -ENwbur /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8/ompi/include/mpif-values.pl /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8_patched/ompi/include/mpif-values.pl
--- /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8/ompi/include/mpif-values.pl	2013-02-27 19:21:39.000000000 +0400
+++ /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8_patched/ompi/include/mpif-values.pl	2013-03-25 12:39:27.000000000 +0400
@@ -299,7 +299,7 @@
 $constants->{MPI_ERR_UNSUPPORTED_DATAREP} = 51;
 $constants->{MPI_ERR_UNSUPPORTED_OPERATION} = 52;
 $constants->{MPI_ERR_WIN} = 53;
-$constants->{MPI_ERR_LASTCODE} = 54;
+$constants->{MPI_ERR_LASTCODE} = 53;
 
 $constants->{MPI_ERR_SYSRESOURCE} = -2;
 
diff -ENwbur /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8/ompi/include/mpi.h.in /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8_patched/ompi/include/mpi.h.in
--- /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8/ompi/include/mpi.h.in	2012-10-05 15:13:10.000000000 +0400
+++ /home/ddontsov/smpi_svn/OpenMPI/openmpi-1.7rc8_patched/ompi/include/mpi.h.in	2013-03-25 12:40:00.000000000 +0400
@@ -562,7 +562,7 @@
 #define MPI_ERR_UNSUPPORTED_DATAREP   51
 #define MPI_ERR_UNSUPPORTED_OPERATION 52
 #define MPI_ERR_WIN                   53
-#define MPI_ERR_LASTCODE              54
+#define MPI_ERR_LASTCODE              53
 
 #define MPI_ERR_SYSRESOURCE          -2
 
