Index: lib/MTT/DoCommand.pm
===================================================================
--- lib/MTT/DoCommand.pm	(revision 1327)
+++ lib/MTT/DoCommand.pm	(working copy)
@@ -619,7 +619,7 @@
         if (FindProgram(qw(padb))) {
 
             my $padb_cmd = "padb --config-option rmgr=mpirun --full-report=$pid";
-            $ret .= "\n $padb_cmd";
+            $ret .= "\n$padb_cmd";
             $ret .= "\n" . `$padb_cmd`;
 
         } else {

