diff -r -U 5 openmpi-1.3.orig/ompi/contrib/vt/vt/tools/vtunify/vt_unify_stats.h openmpi-1.3/ompi/contrib/vt/vt/tools/vtunify/vt_unify_stats.h
--- openmpi-1.3.orig/ompi/contrib/vt/vt/tools/vtunify/vt_unify_stats.h	2009-02-27 08:53:02.000000000 +0000
+++ openmpi-1.3/ompi/contrib/vt/vt/tools/vtunify/vt_unify_stats.h	2009-02-27 09:01:55.000000000 +0000
@@ -17,10 +17,11 @@
 
 #include <algorithm>
 #include <map>
 #include <string>
 #include <vector>
+#include <cctype>
 
 // instance of class Statistics
 class Statistics;
 extern Statistics * theStatistics;
 
@@ -73,13 +74,13 @@
 	 {
 	    std::string _a = funcname, _b = a.funcname;
 	    uint32_t i;
 
 	    for( i = 0; i < funcname.length(); i++ )
-	       _a[i] = tolower( funcname[i] );
+              _a[i] = std::tolower( funcname[i] );
 	    for( i = 0; i < a.funcname.length(); i++ )
-	       _b[i] = tolower( a.funcname[i] );
+              _b[i] = std::tolower( a.funcname[i] );
 	    
 	    return _a < _b;
 	 }
 	 else if( (flags & STAT_SORT_FLAG_FUNCNAME) &&
 		  (flags & STAT_SORT_FLAG_DIR_DOWN ) )

