diff -Naurd openmpi-1.5.a/opal/config/opal_configure_options.m4 openmpi-1.5.b/opal/config/opal_configure_options.m4
--- openmpi-1.5.a/opal/config/opal_configure_options.m4	2011-11-15 19:01:56.000000000 -0600
+++ openmpi-1.5.b/opal/config/opal_configure_options.m4	2011-11-15 19:06:42.000000000 -0600
@@ -480,7 +480,7 @@
 if test "$with_ident_string" = "" -o "$with_ident_string" = "no"; then
     with_ident_string="%VERSION%"
 fi
-with_ident_string="`echo $with_ident_string | sed -e 's/%VERSION%/'$OMPI_VERSION/`"
+with_ident_string="`echo $with_ident_string | sed -e 's/%VERSION%/'"$OMPI_VERSION"'/'`"
 AC_DEFINE_UNQUOTED([OPAL_IDENT_STRING], ["$with_ident_string"],
      [ident string for Open MPI])
 AC_MSG_RESULT([$with_ident_string])

