Index: orte/runtime/orte_mca_params.c
===================================================================
--- orte/runtime/orte_mca_params.c	(revision 27362)
+++ orte/runtime/orte_mca_params.c	(working copy)
@@ -225,7 +225,11 @@
                                 false, false, 1000, &orte_timeout_usec_per_proc);
     
     /* default hostfile */
-    asprintf(&orte_default_hostfile, "%s/openmpi-default-hostfile", opal_install_dirs.sysconfdir);
+    if (NULL != getenv("OPAL_PREFIX")) {
+        asprintf(&orte_default_hostfile, "%s/openmpi-default-hostfile", getenv("OPAL_PREFIX"));
+    } else {
+        asprintf(&orte_default_hostfile, "%s/openmpi-default-hostfile", opal_install_dirs.sysconfdir);
+    }
     mca_base_param_reg_string_name("orte", "default_hostfile",
                                    "Name of the default hostfile (relative or absolute path, \"none\" to ignore environmental or default MCA param setting)",
                                    false, false, orte_default_hostfile, &orte_default_hostfile);

