Index: orte/mca/pls/poe/pls_poe_module.c
===================================================================
--- orte/mca/pls/poe/pls_poe_module.c	(revision 14640)
+++ orte/mca/pls/poe/pls_poe_module.c	(working copy)
@@ -37,6 +37,7 @@
 #include "opal/mca/base/mca_base_param.h"
 #include "opal/util/argv.h"
 #include "opal/util/opal_environ.h"
+#include "opal/util/output.h"
 
 #include "orte/mca/errmgr/errmgr.h"
 #include "orte/mca/gpr/gpr.h"
@@ -69,7 +70,10 @@
 static int pls_poe_signal_job(orte_jobid_t jobid, int32_t signal, opal_list_t *attrs);
 static int pls_poe_signal_proc(const orte_process_name_t *name, int32_t signal);
 static int pls_poe_finalize(void);
-static int pls_poe_cancel_operation(void);
+static int pls_poe_cancel_operation(void) {
+    return ORTE_ERR_NOT_IMPLEMENTED;
+}
+    
 
 orte_pls_base_module_t orte_pls_poe_module = {
     pls_poe_launch_job,

