Index: oshmem/Makefile.am
===================================================================
--- oshmem/Makefile.am	(revision 31810)
+++ oshmem/Makefile.am	(working copy)
@@ -34,6 +34,7 @@
 	include \
 	shmem/c \
 	shmem/fortran
+	shmem/java
 
 if PROJECT_OSHMEM
 # Only traverse these dirs if we're building oshmem
@@ -41,8 +42,7 @@
 	$(MCA_oshmem_FRAMEWORKS_SUBDIRS) \
 	$(MCA_oshmem_FRAMEWORK_COMPONENT_STATIC_SUBDIRS) \
 	. \
-	$(MCA_oshmem_FRAMEWORK_COMPONENT_DSO_SUBDIRS) \
-	shmem/java
+	$(MCA_oshmem_FRAMEWORK_COMPONENT_DSO_SUBDIRS)
 endif
 
 DIST_SUBDIRS = \
