Index: opal/mca/carto/file/carto_file_lex.l
===================================================================
--- opal/mca/carto/file/carto_file_lex.l	(revision 25233)
+++ opal/mca/carto/file/carto_file_lex.l	(working copy)
@@ -1,3 +1,5 @@
+%option nounput
+
 %{ /* -*- C -*- */
 /*
  * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
Index: opal/util/show_help_lex.l
===================================================================
--- opal/util/show_help_lex.l	(revision 25233)
+++ opal/util/show_help_lex.l	(working copy)
@@ -1,3 +1,5 @@
+%option nounput
+
 %{ /* -*- C -*- */
 /*
  * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
Index: opal/util/keyval/keyval_lex.l
===================================================================
--- opal/util/keyval/keyval_lex.l	(revision 25233)
+++ opal/util/keyval/keyval_lex.l	(working copy)
@@ -1,3 +1,5 @@
+%option nounput
+
 %{ /* -*- C -*- */
 /*
  * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
Index: orte/mca/rmaps/rank_file/rmaps_rank_file_lex.l
===================================================================
--- orte/mca/rmaps/rank_file/rmaps_rank_file_lex.l	(revision 25233)
+++ orte/mca/rmaps/rank_file/rmaps_rank_file_lex.l	(working copy)
@@ -1,3 +1,5 @@
+%option nounput
+
 %{ /* -*- C -*- */
 /*
  * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
Index: orte/util/hostfile/hostfile_lex.l
===================================================================
--- orte/util/hostfile/hostfile_lex.l	(revision 25233)
+++ orte/util/hostfile/hostfile_lex.l	(working copy)
@@ -1,3 +1,5 @@
+%option nounput
+
 %{ /* -*- C -*- */
 /*
  * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
