Index: trunk/ompi/mpi/c/testall.c
===================================================================
--- trunk/ompi/mpi/c/testall.c	(revision 26147)
+++ trunk/ompi/mpi/c/testall.c	(working copy)
@@ -67,6 +67,7 @@
     }
 
     if (OPAL_UNLIKELY(0 == count)) {
+        *flag = true;
         return MPI_SUCCESS;
     }
 
Index: trunk/ompi/mpi/c/waitany.c
===================================================================
--- trunk/ompi/mpi/c/waitany.c	(revision 26147)
+++ trunk/ompi/mpi/c/waitany.c	(working copy)
@@ -67,6 +67,7 @@
     }
 
     if (OPAL_UNLIKELY(0 == count)) {
+        *indx = MPI_UNDEFINED;
         return MPI_SUCCESS;
     }
 
Index: trunk/ompi/mpi/c/testany.c
===================================================================
--- trunk/ompi/mpi/c/testany.c	(revision 26147)
+++ trunk/ompi/mpi/c/testany.c	(working copy)
@@ -67,6 +67,8 @@
     }
 
     if (OPAL_UNLIKELY(0 == count)) {
+        *completed = true;
+        *indx = MPI_UNDEFINED;
         return MPI_SUCCESS;
     }
 
Index: trunk/ompi/mpi/c/waitsome.c
===================================================================
--- trunk/ompi/mpi/c/waitsome.c	(revision 26147)
+++ trunk/ompi/mpi/c/waitsome.c	(working copy)
@@ -69,6 +69,7 @@
     }
 
     if (OPAL_UNLIKELY(0 == incount)) {
+        *outcount = MPI_UNDEFINED;
         return MPI_SUCCESS;
     }
 
Index: trunk/ompi/mpi/c/testsome.c
===================================================================
--- trunk/ompi/mpi/c/testsome.c	(revision 26147)
+++ trunk/ompi/mpi/c/testsome.c	(working copy)
@@ -69,6 +69,7 @@
     }
 
     if (OPAL_UNLIKELY(0 == incount)) {
+        *outcount = MPI_UNDEFINED;
         return OMPI_SUCCESS;
     }
 

