Index: orte/orted/pmix/pmix_server_sendrecv.c
===================================================================
--- orte/orted/pmix/pmix_server_sendrecv.c	(revision 32577)
+++ orte/orted/pmix/pmix_server_sendrecv.c	(working copy)
@@ -14,6 +14,8 @@
  * Copyright (c) 2009      Cisco Systems, Inc.  All rights reserved.
  * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
  * Copyright (c) 2013-2014 Intel, Inc.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -673,6 +675,7 @@
                 proc->exit_code = ret;
                 ORTE_FLAG_SET(proc, ORTE_PROC_FLAG_ABORT);
                 ORTE_UPDATE_EXIT_STATUS(ret);
+                ORTE_ACTIVATE_PROC_STATE(&proc->name, ORTE_PROC_STATE_ABORTED);
             }
         }
         /* we will let the ODLS report this to errmgr when the proc exits, so

