diff -r 686ec286164a ompi/communicator/communicator.h
--- a/ompi/communicator/communicator.h	Tue Sep 08 14:39:03 2009 +0200
+++ b/ompi/communicator/communicator.h	Tue Sep 08 15:48:06 2009 +0200
@@ -313,7 +313,7 @@
 {
 #if OPAL_ENABLE_DEBUG
     if(peer_id >= comm->c_remote_group->grp_proc_count) {
-        opal_output(0, "ompi_comm_lookup_peer: invalid peer index (%d)", peer_id);
+        opal_output(0, "ompi_comm_peer_lookup: invalid peer index (%d)", peer_id);
         return (struct ompi_proc_t *) NULL;
     }
 #endif

