diff -up ompi/mca/osc/sm/osc_sm_component.c.orig ompi/mca/osc/sm/osc_sm_component.c
--- ompi/mca/osc/sm/osc_sm_component.c.orig	2014-04-24 10:28:58.790702380 +0300
+++ ompi/mca/osc/sm/osc_sm_component.c	2014-04-24 10:30:15.138137733 +0300
@@ -341,7 +341,7 @@ component_select(struct ompi_win_t *win,
 #if HAVE_PTHREAD_CONDATTR_SETPSHARED && HAVE_PTHREAD_MUTEXATTR_SETPSHARED
         pthread_mutexattr_t mattr;
         pthread_condattr_t cattr;
-        bool blocking_fence;
+        bool blocking_fence = false;
         int flag;
 
         if (OMPI_SUCCESS != ompi_info_get_bool(info, "blocking_fence",
@@ -349,7 +349,7 @@ component_select(struct ompi_win_t *win,
             goto error;
         }
 
-        if (blocking_fence) {
+        if (flag && blocking_fence) {
             ret = pthread_mutexattr_init(&mattr);
             ret = pthread_mutexattr_setpshared(&mattr, PTHREAD_PROCESS_SHARED);
             if (ret != 0) {
