Index: opal/include/opal/sys/ia32/atomic.h
===================================================================
--- opal/include/opal/sys/ia32/atomic.h	(revision 16997)
+++ opal/include/opal/sys/ia32/atomic.h	(working copy)
@@ -10,6 +10,7 @@
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
  * Copyright (c) 2007      Sun Microsystems, Inc.  All rights reserverd.
+ * Copyright (c) 2007      Cisco Systems, Inc.  All rights reserverd.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -51,6 +52,21 @@
 #undef OPAL_HAVE_INLINE_ATOMIC_CMPSET_64
 #define OPAL_HAVE_INLINE_ATOMIC_CMPSET_64 0
 
+/* If we don't have GCC inline assembly, then nothing is inline */
+#if !OMPI_GCC_INLINE_ASSEMBLY
+#undef OPAL_HAVE_INLINE_ATOMIC_MEM_BARRIER
+#define OPAL_HAVE_INLINE_ATOMIC_MEM_BARRIER 0
+
+#undef OPAL_HAVE_INLINE_ATOMIC_CMPSET_32
+#define OPAL_HAVE_INLINE_ATOMIC_CMPSET_32 0
+
+#undef OPAL_HAVE_INLINE_ATOMIC_ADD_32
+#define OPAL_HAVE_INLINE_ATOMIC_ADD_32 0
+
+#undef OPAL_HAVE_INLINE_ATOMIC_SUB_32
+#define OPAL_HAVE_INLINE_ATOMIC_SUB_32 0
+#endif
+
 /**********************************************************************
  *
  * Memory Barriers
Index: opal/include/opal/sys/amd64/atomic.h
===================================================================
--- opal/include/opal/sys/amd64/atomic.h	(revision 16997)
+++ opal/include/opal/sys/amd64/atomic.h	(working copy)
@@ -10,6 +10,7 @@
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
  * Copyright (c) 2007      Sun Microsystems, Inc.  All rights reserverd.
+ * Copyright (c) 2007      Cisco Systems, Inc.  All rights reserverd.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -44,7 +45,18 @@
 
 #define OPAL_HAVE_ATOMIC_CMPSET_64 1
 
+/* If we don't have GCC inline assembly, then nothing is inline */
+#if !OMPI_GCC_INLINE_ASSEMBLY
+#undef OPAL_HAVE_INLINE_ATOMIC_CMPSET_32
+#define OPAL_HAVE_INLINE_ATOMIC_CMPSET_32 0
 
+#undef OPAL_HAVE_INLINE_ATOMIC_CMPSET_64
+#define OPAL_HAVE_INLINE_ATOMIC_CMPSET_64 0
+
+#undef OPAL_HAVE_INLINE_ATOMIC_MEM_BARRIER
+#define OPAL_HAVE_INLINE_ATOMIC_MEM_BARRIER 0
+#endif
+
 /**********************************************************************
  *
  * Memory Barriers

