Index: mpiruntime.h
===================================================================
--- mpiruntime.h	(Revision 15949)
+++ mpiruntime.h	(Arbeitskopie)
@@ -93,7 +93,7 @@
  * Abort the processes of comm
  */
 int ompi_mpi_abort(struct ompi_communicator_t* comm,
-                   int errcode, bool kill_remote_of_intercomm);
+                   int errcode, bool kill_remote_of_intercomm) __opal_attribute_noreturn__;
 
 /**
  * Wait for a TotalView-like debugger if asked.
Index: ompi_mpi_abort.c
===================================================================
--- ompi_mpi_abort.c	(Revision 15949)
+++ ompi_mpi_abort.c	(Arbeitskopie)
@@ -59,7 +59,7 @@
 
     /* Protection for recursive invocation */
     if (have_been_invoked) {
-        return OMPI_SUCCESS;
+        exit (errcode);
     }
     have_been_invoked = true;
 
@@ -193,6 +193,7 @@
 
     /* now that we've aborted everyone else, gracefully die. */
     orte_errmgr.error_detected(errcode, NULL);
-    
-    return OMPI_SUCCESS;
+
+    /* WE SHOULD NEVER GET HERE */
+    exit(errcode);
 }

