diff --git a/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/libnl_utils_common.c b/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/libnl_utils_common.c
index c38a817..fe1e509 100644
--- a/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/libnl_utils_common.c
+++ b/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/libnl_utils_common.c
@@ -39,7 +39,9 @@
  *
  *
  */
+#ifdef HAVE_CONFIG_H
 #include "config.h"
+#endif /* HAVE_CONFIG_H */
 
 #include <errno.h>
 #include <arpa/inet.h>
diff --git a/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/usnic_ip_utils.c b/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/usnic_ip_utils.c
index 815417f..d1727dd 100644
--- a/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/usnic_ip_utils.c
+++ b/opal/mca/common/libfabric/libfabric/prov/usnic/src/usnic_direct/usnic_ip_utils.c
@@ -39,7 +39,9 @@
  *
  *
  */
+#ifdef HAVE_CONFIG_H
 #include "config.h"
+#endif /* HAVE_CONFIG_H */
  
 #include <stdlib.h>
 #include <stdio.h>

