Index: ibm/dynamic/intercomm_create.c
===================================================================
--- ibm/dynamic/intercomm_create.c	(revision 2370)
+++ ibm/dynamic/intercomm_create.c	(working copy)
@@ -104,6 +104,10 @@
     err = MPI_Barrier(abc_intra);
     printf( "%s: barrier (%d)\n", whoami, err );
 
+    MPI_Comm_free(&abc_intra);
+    MPI_Comm_free(&ab_c_inter);
+    MPI_Comm_free(&ab_intra);
+    MPI_Comm_free(&ac_intra);
     MPI_Comm_disconnect(&ab_inter);
     MPI_Comm_disconnect(&ac_inter);
 }
