--- origsrc/openmpi-1.6.3/ompi/debuggers/Makefile.am	2012-04-03 09:30:25.000000000 -0500
+++ src/openmpi-1.6.3/ompi/debuggers/Makefile.am	2012-11-15 19:52:47.037390400 -0600
@@ -58,9 +58,9 @@ libdebuggers_la_SOURCES = \
         $(headers) \
         ompi_debuggers.c
 libdebuggers_la_CPPFLAGS = \
-        -DOMPI_MSGQ_DLL=\"$(pkglibdir)/libompi_dbg_msgq.so\" \
-        -DOMPI_MSGQ_DLL_PREFIX=\"libompi_dbg_msgq\" \
-        -DOMPI_MPIHANDLES_DLL_PREFIX=\"libompi_dbg_mpihandles\"
+        -DOMPI_MSGQ_DLL=\"$(pkglibdir)/cygompi_dbg_msgq.dll\" \
+        -DOMPI_MSGQ_DLL_PREFIX=\"cygompi_dbg_msgq\" \
+        -DOMPI_MPIHANDLES_DLL_PREFIX=\"cygompi_dbg_mpihandles\"
 
 libompi_debugger_canary_la_SOURCES = \
         ompi_debugger_canary.c

