diff --git a/config/ompi_check_openfabrics.m4 b/config/ompi_check_openfabrics.m4
index bb572ae..632df02 100644
--- a/config/ompi_check_openfabrics.m4
+++ b/config/ompi_check_openfabrics.m4
@@ -170,16 +170,17 @@ AC_DEFUN([OMPI_CHECK_OPENFABRICS],[
                $1_have_xrc=1
                AC_CHECK_FUNCS([ibv_create_xrc_rcv_qp],
                               [], [$1_have_xrc=0])
-               AC_CHECK_DECLS([IBV_SRQT_XRC],
-                              [], [$1_have_xrc=0],
-                              [#include <infiniband/verbs.h>])
            fi
            if test "$enable_connectx_xrc" = "yes" \
                && test $$1_have_xrc -eq 1; then
-               AC_CHECK_FUNCS([ibv_cmd_open_xrcd], [$1_have_xrc_domains=1])
+               $1_have_xrc_domains=1
+               AC_CHECK_DECLS([IBV_SRQT_XRC],
+                              [AC_CHECK_FUNCS([ibv_cmd_open_xrcd], 
+                                              [], [$1_have_xrc_domains=0])],
+                              [$1_have_xrc_domains=0],
+                              [#include <infiniband/verbs.h>])
            fi
 
-
            if test "no" != "$enable_openib_dynamic_sl"; then
                # We need ib_types.h file, which is installed with opensm-devel
                # package. However, ib_types.h has a bad include directive,

