Index: orte/tools/orte-restart/orte-restart.c
===================================================================
--- orte/tools/orte-restart/orte-restart.c	(revision 20304)
+++ orte/tools/orte-restart/orte-restart.c	(working copy)
@@ -519,7 +519,7 @@
         goto cleanup;
     }
     if( NULL != orte_restart_globals.hostfile ) {
-        if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, "--hostfile")) ) {
+        if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, "--default-hostfile")) ) {
             exit_status = ret;
             goto cleanup;
         }

