diff -up openmpi-1.7.4rc3/opal/class/opal_hotel.h.includes openmpi-1.7.4rc3/opal/class/opal_hotel.h
--- openmpi-1.7.4rc3/opal/class/opal_hotel.h.includes	2013-09-26 12:22:54.000000000 -0600
+++ openmpi-1.7.4rc3/opal/class/opal_hotel.h	2014-02-03 13:09:22.542594690 -0700
@@ -54,6 +54,7 @@
 
 #include "opal/class/opal_object.h"
 #include "opal/mca/event/event.h"
+#include "opal/prefetch.h"
 
 BEGIN_C_DECLS
 

