Index: opal/datatype/opal_datatype_unpack.c
===================================================================
--- opal/datatype/opal_datatype_unpack.c	(revision 31678)
+++ opal/datatype/opal_datatype_unpack.c	(working copy)
@@ -171,6 +171,8 @@
     uint32_t i, length, count_desc = 1;
     size_t data_length = opal_datatype_basicDatatypes[pElem->elem.common.type]->size;
 
+// GG
+    *user_buffer += pConvertor->pStack->disp;
     DO_DEBUG( opal_output( 0, "unpack partial data start %lu end %lu data_length %lu user %p\n"
                            "\tbConverted %lu total_length %lu count %d\n",
                            (unsigned long)start_position, (unsigned long)end_position, (unsigned long)data_length, *user_buffer,
@@ -230,6 +232,8 @@
             real_data[i] = saved_data[i];
     }
 #endif
+// GG
+    *user_buffer -= pConvertor->pStack->disp;
     return 0;
 }
 

