Index: ompi/proc/proc.c
===================================================================
--- ompi/proc/proc.c	(revision 31891)
+++ ompi/proc/proc.c	(working copy)
@@ -231,7 +231,7 @@
     kvn.key = strdup(OPAL_DSTORE_LOCALITY);
     kvn.type = OPAL_HWLOC_LOCALITY_T;
     kvn.data.uint16 = locality;
-    ret = opal_dstore.store(opal_dstore_internal, (opal_identifier_t*)&proc, &kvn);
+    ret = opal_dstore.store(opal_dstore_internal, (opal_identifier_t*)&proc->proc_name, &kvn);
     OBJ_DESTRUCT(&kvn);
     /* set the proc's local value as well */
     proc->proc_flags = locality;
