--- orte/runtime/orte_globals.h.orig	2009-01-09 18:55:22.000000000 +0100
+++ orte/runtime/orte_globals.h	2009-03-19 15:44:06.822708734 +0100
@@ -109,11 +109,14 @@
 #define ORTE_UPDATE_EXIT_STATUS(newstatus)                                  \
     do {                                                                    \
         if (0 == orte_exit_status && 0 != newstatus) {                      \
+			if ((newstatus & 0377) == 0)									\
+				orte_exit_status = 111;		          						\
+			else                                                            \
+				orte_exit_status = newstatus;                               \
             OPAL_OUTPUT_VERBOSE((1, orte_debug_output,                      \
                                  "%s:%s(%d) updating exit status to %d",    \
                                  ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),        \
-                                 __FILE__, __LINE__, newstatus));           \
-            orte_exit_status = newstatus;                                   \
+                                 __FILE__, __LINE__, orte_exit_status));    \
         }                                                                   \
     } while(0);
 

