Index: /trunk/orte/mca/ras/base/ras_base_node.c
===================================================================
--- /trunk/orte/mca/ras/base/ras_base_node.c (revision 20741)
+++ /trunk/orte/mca/ras/base/ras_base_node.c (revision 20856)
@@ -97,5 +97,6 @@
          * if this node is the same as the HNP's node so we don't double-enter it
          */
-        if (opal_ifislocal(node->name)) {
+        if (0 == strcmp(node->name, hnp_node->name) ||
+            opal_ifislocal(node->name)) {
             OPAL_OUTPUT_VERBOSE((5, orte_ras_base.ras_output,
                                  "%s ras:base:node_insert updating HNP info to %ld slots",

