diff --git a/opal/mca/btl/openib/btl_openib_component.c b/opal/mca/btl/openib/btl_openib_component.c
index 829a77d..9d57e72 100644
--- a/opal/mca/btl/openib/btl_openib_component.c
+++ b/opal/mca/btl/openib/btl_openib_component.c
@@ -982,14 +982,14 @@ static void device_destruct(mca_btl_openib_device_t *device)
     }
 #endif
 
-    if (ibv_dealloc_pd(device->ib_pd)) {
+    if (NULL != device->ib_pd && ibv_dealloc_pd(device->ib_pd)) {
         BTL_VERBOSE(("Warning! Failed to release PD"));
         goto device_error;
     }
 
     OBJ_DESTRUCT(&device->device_lock);
 
-    if (ibv_close_device(device->ib_dev_context)) {
+    if (NULL != device->ib_dev_context && ibv_close_device(device->ib_dev_context)) {
         if (1 == opal_leave_pinned || opal_leave_pinned_pipeline) {
             BTL_VERBOSE(("Warning! Failed to close device"));
             goto device_error;
@@ -2237,6 +2237,7 @@ error:
 
     if (device->ib_pd) {
         ibv_dealloc_pd(device->ib_pd);
+        device->ib_pd = NULL;
     }
 
     if (OPAL_SUCCESS != ret) {
@@ -2248,6 +2249,7 @@ error:
 
     if (device->ib_dev_context) {
         ibv_close_device(device->ib_dev_context);
+        device->ib_dev_context = NULL;
     }
     if (NULL != allowed_ports) {
         free(allowed_ports);

