Index: orte/mca/plm/base/plm_base_launch_support.c
===================================================================
--- orte/mca/plm/base/plm_base_launch_support.c	(revision 32641)
+++ orte/mca/plm/base/plm_base_launch_support.c	(working copy)
@@ -1170,17 +1170,21 @@
         rml_uri = orte_rml.get_contact_info();
     } else {
         rml_uri = orte_rml.get_contact_info();
+        (void)asprintf(&param, "\"%s\"", rml_uri);
         opal_argv_append(argc, argv, "-mca");
         opal_argv_append(argc, argv, "orte_parent_uri");
-        opal_argv_append(argc, argv, rml_uri);
+        opal_argv_append(argc, argv, param);
         free(rml_uri);
+        free(param);
     
         rml_uri = strdup(orte_process_info.my_hnp_uri);
     }
+    (void)asprintf(&param, "\"%s\"", rml_uri);
     opal_argv_append(argc, argv, "-mca");
     opal_argv_append(argc, argv, "orte_hnp_uri");
-    opal_argv_append(argc, argv, rml_uri);
+    opal_argv_append(argc, argv, param);
     free(rml_uri);
+    free(param);
 
     /* if we have static ports, pass the node list */
     if (orte_static_ports && NULL != nodes) {
Index: orte/orted/orted_main.c
===================================================================
--- orte/orted/orted_main.c	(revision 32641)
+++ orte/orted/orted_main.c	(working copy)
@@ -682,7 +682,11 @@
                                   &orte_parent_uri);
     if (NULL != orte_parent_uri) {
         orte_process_name_t parent;
-
+        if ('"' == orte_parent_uri[0]) {
+            /* if the first char is a quote, then so will the last one be */
+            orte_parent_uri[strlen(orte_parent_uri)-1] = '\0';
+            memmove(orte_parent_uri, orte_parent_uri + 1, strlen(orte_parent_uri) - 1);
+        }
         /* set the contact info into the hash table */
         orte_rml.set_contact_info(orte_parent_uri);
         ret = orte_rml_base_parse_uris(orte_parent_uri, &parent, NULL);

