diff -r 4bd1b2ed9861 ompi/mca/btl/openib/btl_openib_component.c
--- a/ompi/mca/btl/openib/btl_openib_component.c	Tue May 08 21:49:05 2012 +0000
+++ b/ompi/mca/btl/openib/btl_openib_component.c	Wed May 09 15:45:18 2012 +0200
@@ -877,8 +877,8 @@ static int init_one_port(opal_list_t *bt
                     openib_btl->super.btl_bandwidth = 8000;
                     break;
                 case 8:
-                    /* 40.0Gbps * 0.8, in megabits */
-                    openib_btl->super.btl_bandwidth = 32000;
+                    /* FDR-10 10.0Gbps * 0.8, in megabits */
+                    openib_btl->super.btl_bandwidth = 8000;
                     break;
                 default:
                     /* Who knows?  Declare this port unreachable (do

