Index: Makefile
===================================================================
--- Makefile	(revision 25980)
+++ Makefile	(working copy)
@@ -49,19 +49,19 @@
 # if Open MPI was build with the relevant language bindings.
 
 all: hello_c ring_c connectivity_c
-	@ if test "`ompi_info --parsable | grep bindings:cxx:yes`" != ""; then \
+	@ if test "`ompi_info --parsable | grep bindings:cxx:yes || true`" != ""; then \
 	    $(MAKE) hello_cxx ring_cxx; \
 	fi
-	@ if test "`ompi_info --parsable | grep bindings:f77:yes`" != ""; then \
+	@ if test "`ompi_info --parsable | grep bindings:f77:yes || true`" != ""; then \
 	    $(MAKE) hello_f77 ring_f77; \
 	fi
-	@ if test "`ompi_info --parsable | grep bindings:f90:yes`" != ""; then \
+	@ if test "`ompi_info --parsable | grep bindings:f90:yes || true`" != ""; then \
 	    $(MAKE) hello_f90 ring_f90; \
 	fi
-	@ if test "`ompi_info --parsable | grep bindings:java:yes`" != ""; then \
+	@ if test "`ompi_info --parsable | grep bindings:java:yes || true``" != ""; then \
 	    $(MAKE) Hello.class; \
 	fi
-	@ if test "`ompi_info --parsable | grep bindings:java:yes`" != ""; then \
+	@ if test "`ompi_info --parsable | grep bindings:java:yes || true`" != ""; then \
 	    $(MAKE) Ring.class; \
 	fi
 

