Index: ompi/mpi/c/comm_disconnect.c
===================================================================
--- ompi/mpi/c/comm_disconnect.c	(revision 32594)
+++ ompi/mpi/c/comm_disconnect.c	(working copy)
@@ -10,6 +10,8 @@
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
  * Copyright (c) 2006      Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -61,14 +63,13 @@
 
     OPAL_CR_ENTER_LIBRARY();
 
+    /* Always invoke coll_barrier in order to avoid a race condition in pmix */
+    (*comm)->c_coll.coll_barrier(*comm, (*comm)->c_coll.coll_barrier_module);
     if ( OMPI_COMM_IS_DYNAMIC(*comm)) {
         if (OMPI_SUCCESS != ompi_dpm.disconnect (*comm)) {
             ret = OMPI_ERRHANDLER_INVOKE(MPI_COMM_WORLD, MPI_ERR_COMM, FUNC_NAME);
         }
     }
-    else {
-        (*comm)->c_coll.coll_barrier(*comm, (*comm)->c_coll.coll_barrier_module);
-    }
 
     ompi_comm_free(comm);
 

