--- openmpi-1.2b4/orte/mca/pls/rsh/pls_rsh_module.c~	2007-02-16 19:20:49.000000000 +0100
+++ openmpi-1.2b4/orte/mca/pls/rsh/pls_rsh_module.c	2007-03-01 11:12:48.000000000 +0100
@@ -115,6 +115,7 @@
 
 enum {
     ORTE_PLS_RSH_SHELL_BASH = 0,
+    ORTE_PLS_RSH_SHELL_ZSH,
     ORTE_PLS_RSH_SHELL_TCSH,
     ORTE_PLS_RSH_SHELL_CSH,
     ORTE_PLS_RSH_SHELL_KSH,
@@ -126,6 +127,7 @@
 
 static const char * orte_pls_rsh_shell_name[] = {
     "bash",
+    "zsh",
     "tcsh",       /* tcsh has to be first otherwise strstr finds csh */
     "csh",
     "ksh",
@@ -576,6 +578,7 @@
                 switch (i) {
                 case ORTE_PLS_RSH_SHELL_SH:  /* fall through */
                 case ORTE_PLS_RSH_SHELL_KSH: /* fall through */
+                case ORTE_PLS_RSH_SHELL_ZSH: /* fall through */
                 case ORTE_PLS_RSH_SHELL_BASH: local_sh = true; break;
                 case ORTE_PLS_RSH_SHELL_TCSH: /* fall through */
                 case ORTE_PLS_RSH_SHELL_CSH:  local_csh = true; break;
