diff --git a/ompi/mpi/java/c/mpi_MPI.c b/ompi/mpi/java/c/mpi_MPI.c
index 7c3a3ba..219da6e 100644
--- a/ompi/mpi/java/c/mpi_MPI.c
+++ b/ompi/mpi/java/c/mpi_MPI.c
@@ -62,6 +62,7 @@
 #include <sys/stat.h>
 #endif
 #include <dlfcn.h>
+#include <poll.h>
 
 #include "opal/util/output.h"
 #include "opal/datatype/opal_convertor.h"
@@ -121,6 +122,11 @@ OBJ_CLASS_INSTANCE(ompi_java_buffer_t,
  */
 jint JNI_OnLoad(JavaVM *vm, void *reserved)
 {
+    char *env = getenv("OMPI_ATTACH");
+    if (NULL != env && 0 < atoi(env)) {
+        volatile int _dbg = 1;
+        while (_dbg) poll(NULL, 0, 1);
+    }
     libmpi = dlopen("libmpi." OPAL_DYN_LIB_SUFFIX, RTLD_NOW | RTLD_GLOBAL);
 
     if(libmpi == NULL)

