--- origsrc/openmpi-1.9a1r27710/opal/util/path.c	2012-12-20 03:00:25.000000000 +0100
+++ src/openmpi-1.9a1r27710/opal/util/path.c	2012-12-21 14:34:15.432823000 +0100
@@ -547,7 +547,7 @@
 #if defined(__SVR4) && defined(__sun)
     struct statvfs buf;
 #elif defined(__linux__) || defined (__BSD) ||                                 \
-      (defined(__APPLE__) && defined(__MACH__))
+      (defined(__APPLE__) && defined(__MACH__)) || defined(__CYGWIN__)
     struct statfs buf;
 #endif
 
@@ -560,7 +560,7 @@
 #if defined(__SVR4) && defined(__sun)
         rc = statvfs(path, &buf);
 #elif defined(__linux__) || defined (__BSD) ||                                 \
-      (defined(__APPLE__) && defined(__MACH__))
+      (defined(__APPLE__) && defined(__MACH__)) || defined(__CYGWIN__)
         rc = statfs(path, &buf);
 #endif
         err = errno;

