Index: orte/mca/odls/base/odls_base_default_fns.c
===================================================================
--- orte/mca/odls/base/odls_base_default_fns.c	(revision 32554)
+++ orte/mca/odls/base/odls_base_default_fns.c	(working copy)
@@ -15,6 +15,8 @@
  *                         All rights reserved.
  * Copyright (c) 2011-2013 Cisco Systems, Inc.  All rights reserved.
  * Copyright (c) 2013-2014 Intel, Inc.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -1772,6 +1774,20 @@
                         ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),
                         ORTE_NAME_PRINT(&proc->name), (long)proc->pid);
     
+#if 1
+    /* the child might have died but the SIGCHLD event has not
+     * been processed by libevent yet. try to waitpid in order
+     * to set proc->exit_code */
+    if (ORTE_FLAG_TEST(proc, ORTE_PROC_FLAG_ALIVE)) {
+        int status;
+        /* FIXME : loop as long as errno == EINTR */
+        if (waitpid(proc->pid, &status, WNOHANG) == proc->pid) {
+            int state = ORTE_PROC_STATE_WAITPID_FIRED;
+            proc->exit_code = status;
+            ORTE_ACTIVATE_PROC_STATE(&proc->name, state);
+        }
+    }
+#endif
     /* if the child was previously flagged as dead, then just
      * update its exit status and
      * ensure that its exit state gets reported to avoid hanging

