diff -uNr openmpi-1.3/ompi/contrib/vt/vt/vtlib/vt_otf_gen.c openmpi-1.3_mod/ompi/contrib/vt/vt/vtlib/vt_otf_gen.c
--- openmpi-1.3/ompi/contrib/vt/vt/vtlib/vt_otf_gen.c	2008-11-04 21:56:41.000000000 +0100
+++ openmpi-1.3_mod/ompi/contrib/vt/vt/vtlib/vt_otf_gen.c	2009-01-23 10:25:33.000000000 +0100
@@ -26,6 +26,7 @@
 #include <unistd.h>
 #include <string.h>
 #include <stdlib.h>
+#include <sys/param.h>
 
 #include "otf.h"
 
diff -uNr openmpi-1.3/ompi/contrib/vt/vt/vtlib/vt_thrd.c openmpi-1.3_mod/ompi/contrib/vt/vt/vtlib/vt_thrd.c
--- openmpi-1.3/ompi/contrib/vt/vt/vtlib/vt_thrd.c	2008-11-04 21:56:41.000000000 +0100
+++ openmpi-1.3_mod/ompi/contrib/vt/vt/vtlib/vt_thrd.c	2009-01-23 10:26:10.000000000 +0100
@@ -24,6 +24,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <sys/param.h>
 
 
 #if (defined (VT_MPI) || defined (VT_OMPI))
diff -uNr openmpi-1.3/opal/util/path.c openmpi-1.3_mod/opal/util/path.c
--- openmpi-1.3/opal/util/path.c	2008-11-04 21:56:24.000000000 +0100
+++ openmpi-1.3_mod/opal/util/path.c	2009-01-23 09:36:46.000000000 +0100
@@ -28,6 +28,10 @@
 #ifdef HAVE_SYS_STAT_H
 #include <sys/stat.h>
 #endif
+#ifdef HAVE_SYS_PARAM_H
+#include <sys/param.h>
+#endif
+
 
 #include "opal/util/path.h"
 #include "opal/util/os_path.h"
diff -uNr openmpi-1.3/orte/mca/plm/rsh/plm_rsh_component.c openmpi-1.3_mod/orte/mca/plm/rsh/plm_rsh_component.c
--- openmpi-1.3/orte/mca/plm/rsh/plm_rsh_component.c	2008-11-04 21:56:47.000000000 +0100
+++ openmpi-1.3_mod/orte/mca/plm/rsh/plm_rsh_component.c	2009-01-23 09:52:52.000000000 +0100
@@ -33,6 +33,9 @@
 #include <unistd.h>
 #endif
 #include <ctype.h>
+#ifdef HAVE_SYS_PARAM_H
+#include <sys/param.h>
+#endif
 
 #include "opal/util/opal_environ.h"
 #include "opal/util/argv.h"
diff -uNr openmpi-1.3/orte/tools/orterun/debuggers.c openmpi-1.3_mod/orte/tools/orterun/debuggers.c
--- openmpi-1.3/orte/tools/orterun/debuggers.c	2008-11-04 21:56:48.000000000 +0100
+++ openmpi-1.3_mod/orte/tools/orterun/debuggers.c	2009-01-23 10:00:42.000000000 +0100
@@ -102,6 +102,9 @@
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif  /* HAVE_UNISTD_H */
+#ifdef HAVE_SYS_PARAM_H
+#include <sys/param.h>
+#endif /*HAVE_SYS_PARAM_H*/
 #include <ctype.h>
 
 #include "opal/util/argv.h"
