diff --git a/config/opal_config_pthreads.m4 b/config/opal_config_pthreads.m4
index 7bc0bfe..eb7803c 100644
--- a/config/opal_config_pthreads.m4
+++ b/config/opal_config_pthreads.m4
@@ -178,7 +178,14 @@ AC_DEFUN([OPAL_INTL_POSIX_THREADS_PLAIN_C], [
 #
 if test "$opal_pthread_c_success" = "0"; then
   AC_MSG_CHECKING([if C compiler and POSIX threads work as is])
-
+  case "${host_cpu}-${host-_os}" in
+    *solaris*)
+      if test "`echo $CPPFLAGS | $GREP 'D_REENTRANT'`" = ""; then
+        PTHREAD_CPPFLAGS="-D_REENTRANT"
+        CPPFLAGS="$CPPFLAGS $PTHREAD_CPPFLAGS"
+      fi
+    ;;
+  esac
   AC_LANG_PUSH(C)
   OPAL_INTL_PTHREAD_TRY_LINK(opal_pthread_c_success=1,
                             opal_pthread_c_success=0)
@@ -198,7 +205,14 @@ AC_DEFUN([OPAL_INTL_POSIX_THREADS_PLAIN_CXX], [
 #
 if test "$opal_pthread_cxx_success" = "0"; then
   AC_MSG_CHECKING([if C++ compiler and POSIX threads work as is])
-
+  case "${host_cpu}-${host-_os}" in
+    *solaris*)
+      if test "`echo $CXXCPPFLAGS | $GREP 'D_REENTRANT'`" = ""; then
+        PTHREAD_CXXCPPFLAGS="-D_REENTRANT"
+        CXXCPPFLAGS="$CXXCPPFLAGS $PTHREAD_CXXCPPFLAGS"
+      fi
+    ;;
+  esac
   AC_LANG_PUSH(C++)
   OPAL_INTL_PTHREAD_TRY_LINK(opal_pthread_cxx_success=1, 
                             opal_pthread_cxx_success=0)

