diff --git a/ompi/datatype/ompi_datatype_args.c b/ompi/datatype/ompi_datatype_args.c
index d6a33b3..0b5e01c 100644
--- a/ompi/datatype/ompi_datatype_args.c
+++ b/ompi/datatype/ompi_datatype_args.c
@@ -100,8 +100,8 @@ typedef struct __dt_args {
         if( pArgs->ci == 0 ) pArgs->i = NULL;                           \
         else pArgs->i = (int*)buf;                                      \
         pArgs->ref_count = 1;                                           \
-        pArgs->total_pack_size = (4 + (IC)) * sizeof(int) +             \
-            (AC) * sizeof(OPAL_PTRDIFF_TYPE);                           \
+        pArgs->total_pack_size = (4 + (IC) - (DC)) * sizeof(int) +      \
+            ((AC) + (DC)) * sizeof(OPAL_PTRDIFF_TYPE);                  \
         (PDATA)->args = (void*)pArgs;                                   \
         (PDATA)->packed_description = NULL;                             \
     } while(0)
@@ -494,7 +494,9 @@ int ompi_datatype_get_pack_description( ompi_datatype_t* datatype,
         recursive_buffer = datatype->packed_description;
         __ompi_datatype_pack_description( datatype, &recursive_buffer, &next_index );
         if( !ompi_datatype_is_predefined(datatype) ) {
-            args->total_pack_size = (uintptr_t)((char*)recursive_buffer - (char*)datatype->packed_description);
+            uintptr_t new_size = (uintptr_t)((char*)recursive_buffer - (char*)datatype->packed_description);
+            assert(new_size <= args->total_pack_size);
+            args->total_pack_size = new_size;
         }
     }
 
