diff --git a/ompi/mca/common/verbs_usnic/configure.m4 b/ompi/mca/common/verbs_usnic/configure.m4
index 60aaa62..ad6eeb1 100644
--- a/ompi/mca/common/verbs_usnic/configure.m4
+++ b/ompi/mca/common/verbs_usnic/configure.m4
@@ -40,15 +40,25 @@ AC_DEFUN([MCA_ompi_common_verbs_usnic_CONFIG],[
     AC_CONFIG_FILES([ompi/mca/common/verbs_usnic/Makefile])
     common_verbs_usnic_happy="no"
 
-    OMPI_CHECK_OPENFABRICS([common_verbs_usnic],
-                           [common_verbs_usnic_happy="yes"])
+    AC_ARG_WITH([usnic],
+                [AS_HELP_STRING([--with-usnic],
+                                [If specified, cause an error if usNIC
+                                 support cannot be built])])
 
-    AS_IF([test "$common_verbs_usnic_happy" = "yes"],
-          [$1],
-          [$2])
+    # If --without-usnic was specified, then gracefully exit.
+    # Otherwise, do the rest of the config.
+    AS_IF([test "x$with_usnic" = "xno"],
+          [AC_MSG_WARN([--without-usnic specified; skipping verbs_usnic])
+           $2],
+          [OMPI_CHECK_OPENFABRICS([common_verbs_usnic],
+                                  [common_verbs_usnic_happy="yes"])
 
-    # substitute in the things needed to build openib
-    AC_SUBST([common_verbs_usnic_CPPFLAGS])
-    AC_SUBST([common_verbs_usnic_LDFLAGS])
-    AC_SUBST([common_verbs_usnic_LIBS])
+           AS_IF([test "$common_verbs_usnic_happy" = "yes"],
+                 [$1],
+                 [$2])
+
+           # substitute in the things needed to build openib
+           AC_SUBST([common_verbs_usnic_CPPFLAGS])
+           AC_SUBST([common_verbs_usnic_LDFLAGS])
+           AC_SUBST([common_verbs_usnic_LIBS])])
 ])dnl

