diff --git a/src/topology-linux.c b/src/topology-linux.c
index e1f46cb..99a6381 100644
--- a/src/topology-linux.c
+++ b/src/topology-linux.c
@@ -475,7 +475,7 @@ hwloc_linux_foreach_proc_tid(hwloc_topology_t topology,
   char taskdir_path[128];
   DIR *taskdir;
   pid_t *tids, *newtids;
-  unsigned i, nr, newnr;
+  unsigned i, nr, newnr, failed;
   int err;
 
   if (pid)
@@ -497,11 +497,17 @@ hwloc_linux_foreach_proc_tid(hwloc_topology_t topology,
 
  retry:
   /* apply the callback to all threads */
+  failed=0;
   for(i=0; i<nr; i++) {
     err = cb(topology, tids[i], data, i);
     if (err < 0)
-      goto out_with_tids;
+      failed++;
   }
+  /* some may fail (if threads disappear), but some should succeed.
+   * if all failed, abort with the last errno.
+   */
+  if (failed == nr)
+    goto out_with_tids;
 
   /* re-read the list of thread and retry if it changed in the meantime */
   err = hwloc_linux_get_proc_tids(taskdir, &newnr, &newtids);

