diff -ruN ./contrib/Makefile.am ../openmpi-1.4.2.new/contrib/Makefile.am
--- ./contrib/Makefile.am	2010-04-13 11:30:27.000000000 -0600
+++ ../openmpi-1.4.2.new/contrib/Makefile.am	2010-05-30 08:39:58.000000000 -0600
@@ -100,9 +100,9 @@
         platform/win32/ConfigFiles/mpif77-wrapper-data.txt.cmake \
         platform/win32/ConfigFiles/opal_config.h.cmake \
         platform/win32/ConfigFiles/revision.in \
-        platform/ibm/debug-ppc32-gcc \ 
-        platform/ibm/debug-ppc64-gcc \ 
-        platform/ibm/optimized-ppc32-gcc \ 
-        platform/ibm/optimized-ppc64-gcc 
+        platform/ibm/debug-ppc32-gcc \
+        platform/ibm/debug-ppc64-gcc \
+        platform/ibm/optimized-ppc32-gcc \
+        platform/ibm/optimized-ppc64-gcc
 
 dist_pkgdata_DATA = openmpi-valgrind.supp
diff -ruN ./contrib/Makefile.in ../openmpi-1.4.2.new/contrib/Makefile.in
--- ./contrib/Makefile.in	2010-05-04 10:06:10.000000000 -0600
+++ ../openmpi-1.4.2.new/contrib/Makefile.in	2010-05-30 08:40:31.000000000 -0600
@@ -743,6 +743,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -1003,7 +1004,7 @@
         platform/ibm/debug-ppc32-gcc \
         platform/ibm/debug-ppc64-gcc \
         platform/ibm/optimized-ppc32-gcc \
-        platform/ibm/optimized-ppc64-gcc 
+        platform/ibm/optimized-ppc64-gcc
 
 dist_pkgdata_DATA = openmpi-valgrind.supp
 all: all-am
diff -ruN ./ompi/tools/ompi_info/Makefile.am ../openmpi-1.4.2.new/ompi/tools/ompi_info/Makefile.am
--- ./ompi/tools/ompi_info/Makefile.am	2009-12-08 13:36:19.000000000 -0700
+++ ../openmpi-1.4.2.new/ompi/tools/ompi_info/Makefile.am	2010-05-30 08:58:40.000000000 -0600
@@ -66,7 +66,7 @@
         output.cc \
         param.cc \
         version.cc 
-ompi_info_LDADD = $(top_builddir)/ompi/libmpi.la
+ompi_info_LDADD = $(top_builddir)/ompi/libmpi.la $(OMPI_WRAPPER_EXTRA_LDFLAGS) $(OMPI_WRAPPER_EXTRA_LIBS)
 
 clean-local:
 	test -z "$(OMPI_CXX_TEMPLATE_REPOSITORY)" || $(RM) -rf $(OMPI_CXX_TEMPLATE_REPOSITORY)
diff -ruN ./ompi/tools/ompi_info/Makefile.in ../openmpi-1.4.2.new/ompi/tools/ompi_info/Makefile.in
--- ./ompi/tools/ompi_info/Makefile.in	2010-05-04 10:06:50.000000000 -0600
+++ ../openmpi-1.4.2.new/ompi/tools/ompi_info/Makefile.in	2010-05-30 08:59:17.000000000 -0600
@@ -254,7 +254,9 @@
 am_ompi_info_OBJECTS = components.$(OBJEXT) ompi_info.$(OBJEXT) \
 	output.$(OBJEXT) param.$(OBJEXT) version.$(OBJEXT)
 ompi_info_OBJECTS = $(am_ompi_info_OBJECTS)
-ompi_info_DEPENDENCIES = $(top_builddir)/ompi/libmpi.la
+am__DEPENDENCIES_1 =
+ompi_info_DEPENDENCIES = $(top_builddir)/ompi/libmpi.la \
+	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1)
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)/opal/include -I$(top_builddir)/orte/include -I$(top_builddir)/ompi/include -I$(top_builddir)/opal/mca/paffinity/linux/plpa/src/libplpa
 depcomp = $(SHELL) $(top_srcdir)/config/depcomp
 am__depfiles_maybe = depfiles
@@ -785,6 +787,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -997,7 +1000,7 @@
         param.cc \
         version.cc 
 
-ompi_info_LDADD = $(top_builddir)/ompi/libmpi.la
+ompi_info_LDADD = $(top_builddir)/ompi/libmpi.la $(OMPI_WRAPPER_EXTRA_LDFLAGS) $(OMPI_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:
diff -ruN ./ompi/tools/ompi-server/Makefile.am ../openmpi-1.4.2.new/ompi/tools/ompi-server/Makefile.am
--- ./ompi/tools/ompi-server/Makefile.am	2009-12-08 13:36:19.000000000 -0700
+++ ../openmpi-1.4.2.new/ompi/tools/ompi-server/Makefile.am	2010-05-30 08:59:38.000000000 -0600
@@ -41,7 +41,7 @@
 endif # OMPI_INSTALL_BINARIES
 
 ompi_server_SOURCES = ompi-server.c 
-ompi_server_LDADD = $(top_builddir)/ompi/libmpi.la
+ompi_server_LDADD = $(top_builddir)/ompi/libmpi.la $(OMPI_WRAPPER_EXTRA_LDFLAGS) $(OMPI_WRAPPER_EXTRA_LIBS)
 
 distclean-local:
 	rm -f $(man_pages)
diff -ruN ./ompi/tools/ompi-server/Makefile.in ../openmpi-1.4.2.new/ompi/tools/ompi-server/Makefile.in
--- ./ompi/tools/ompi-server/Makefile.in	2010-05-04 10:06:50.000000000 -0600
+++ ../openmpi-1.4.2.new/ompi/tools/ompi-server/Makefile.in	2010-05-30 09:00:01.000000000 -0600
@@ -253,7 +253,9 @@
 PROGRAMS = $(bin_PROGRAMS)
 am_ompi_server_OBJECTS = ompi-server.$(OBJEXT)
 ompi_server_OBJECTS = $(am_ompi_server_OBJECTS)
-ompi_server_DEPENDENCIES = $(top_builddir)/ompi/libmpi.la
+am__DEPENDENCIES_1 =
+ompi_server_DEPENDENCIES = $(top_builddir)/ompi/libmpi.la \
+	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1)
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)/opal/include -I$(top_builddir)/orte/include -I$(top_builddir)/ompi/include -I$(top_builddir)/opal/mca/paffinity/linux/plpa/src/libplpa
 depcomp = $(SHELL) $(top_srcdir)/config/depcomp
 am__depfiles_maybe = depfiles
@@ -775,6 +777,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -959,7 +962,7 @@
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@dist_pkgdata_DATA = help-ompi-server.txt
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@nodist_man_MANS = $(man_pages)
 ompi_server_SOURCES = ompi-server.c 
-ompi_server_LDADD = $(top_builddir)/ompi/libmpi.la
+ompi_server_LDADD = $(top_builddir)/ompi/libmpi.la $(OMPI_WRAPPER_EXTRA_LDFLAGS) $(OMPI_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:
diff -ruN ./orte/tools/orte-clean/Makefile.am ../openmpi-1.4.2.new/orte/tools/orte-clean/Makefile.am
--- ./orte/tools/orte-clean/Makefile.am	2009-12-08 13:36:41.000000000 -0700
+++ ../openmpi-1.4.2.new/orte/tools/orte-clean/Makefile.am	2010-05-30 08:39:04.000000000 -0600
@@ -40,7 +40,7 @@
 endif # OMPI_INSTALL_BINARIES
 
 orte_clean_SOURCES = orte-clean.c
-orte_clean_LDADD = $(top_builddir)/orte/libopen-rte.la
+orte_clean_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 
 endif # !ORTE_DISABLE_FULL_SUPPORT
 
diff -ruN ./orte/tools/orte-clean/Makefile.in ../openmpi-1.4.2.new/orte/tools/orte-clean/Makefile.in
--- ./orte/tools/orte-clean/Makefile.in	2010-05-04 10:08:01.000000000 -0600
+++ ../openmpi-1.4.2.new/orte/tools/orte-clean/Makefile.in	2010-05-30 08:39:25.000000000 -0600
@@ -255,7 +255,10 @@
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@am_orte_clean_OBJECTS =  \
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@	orte-clean.$(OBJEXT)
 orte_clean_OBJECTS = $(am_orte_clean_OBJECTS)
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_clean_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la
+am__DEPENDENCIES_1 =
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_clean_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1) \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1)
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)/opal/include -I$(top_builddir)/orte/include -I$(top_builddir)/ompi/include -I$(top_builddir)/opal/mca/paffinity/linux/plpa/src/libplpa
 depcomp = $(SHELL) $(top_srcdir)/config/depcomp
 am__depfiles_maybe = depfiles
@@ -777,6 +780,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -961,7 +965,7 @@
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@nodist_man_MANS = $(man_pages)
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@dist_pkgdata_DATA = help-orte-clean.txt
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_clean_SOURCES = orte-clean.c
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_clean_LDADD = $(top_builddir)/orte/libopen-rte.la
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_clean_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:
diff -ruN ./orte/tools/orted/Makefile.am ../openmpi-1.4.2.new/orte/tools/orted/Makefile.am
--- ./orte/tools/orted/Makefile.am	2009-12-08 13:36:41.000000000 -0700
+++ ../openmpi-1.4.2.new/orte/tools/orted/Makefile.am	2010-05-30 08:48:02.000000000 -0600
@@ -45,7 +45,7 @@
 #  nice for systems that don't have all the shared 
 #  libraries on the computes 
 orted_LDFLAGS =
-orted_LDADD = $(top_builddir)/orte/libopen-rte.la
+orted_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 
 endif # !ORTE_DISABLE_FULL_SUPPORT
 
diff -ruN ./orte/tools/orted/Makefile.in ../openmpi-1.4.2.new/orte/tools/orted/Makefile.in
--- ./orte/tools/orted/Makefile.in	2010-05-04 10:08:03.000000000 -0600
+++ ../openmpi-1.4.2.new/orte/tools/orted/Makefile.in	2010-05-30 08:48:13.000000000 -0600
@@ -252,7 +252,10 @@
 am__orted_SOURCES_DIST = orted.c
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@am_orted_OBJECTS = orted.$(OBJEXT)
 orted_OBJECTS = $(am_orted_OBJECTS)
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orted_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la
+am__DEPENDENCIES_1 =
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orted_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1) \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1)
 orted_LINK = $(LIBTOOL) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) \
 	--mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(orted_LDFLAGS) \
 	$(LDFLAGS) -o $@
@@ -768,6 +771,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -958,7 +962,7 @@
 #  nice for systems that don't have all the shared 
 #  libraries on the computes 
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@orted_LDFLAGS = 
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orted_LDADD = $(top_builddir)/orte/libopen-rte.la
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orted_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:
diff -ruN ./orte/tools/orte-iof/Makefile.am ../openmpi-1.4.2.new/orte/tools/orte-iof/Makefile.am
--- ./orte/tools/orte-iof/Makefile.am	2009-12-08 13:36:41.000000000 -0700
+++ ../openmpi-1.4.2.new/orte/tools/orte-iof/Makefile.am	2010-05-30 08:40:24.000000000 -0600
@@ -40,7 +40,7 @@
 endif # OMPI_INSTALL_BINARIES
 
 orte_iof_SOURCES = orte-iof.c
-orte_iof_LDADD = $(top_builddir)/orte/libopen-rte.la
+orte_iof_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 
 endif # !ORTE_DISABLE_FULL_SUPPORT
 
diff -ruN ./orte/tools/orte-iof/Makefile.in ../openmpi-1.4.2.new/orte/tools/orte-iof/Makefile.in
--- ./orte/tools/orte-iof/Makefile.in	2010-05-04 10:08:01.000000000 -0600
+++ ../openmpi-1.4.2.new/orte/tools/orte-iof/Makefile.in	2010-05-30 08:40:39.000000000 -0600
@@ -255,7 +255,10 @@
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@am_orte_iof_OBJECTS =  \
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@	orte-iof.$(OBJEXT)
 orte_iof_OBJECTS = $(am_orte_iof_OBJECTS)
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_iof_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la
+am__DEPENDENCIES_1 =
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_iof_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1) \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1)
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)/opal/include -I$(top_builddir)/orte/include -I$(top_builddir)/ompi/include -I$(top_builddir)/opal/mca/paffinity/linux/plpa/src/libplpa
 depcomp = $(SHELL) $(top_srcdir)/config/depcomp
 am__depfiles_maybe = depfiles
@@ -777,6 +780,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -961,7 +965,7 @@
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@nodist_man_MANS = $(man_pages)
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@dist_pkgdata_DATA = help-orte-iof.txt
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_iof_SOURCES = orte-iof.c
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_iof_LDADD = $(top_builddir)/orte/libopen-rte.la
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_iof_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:
diff -ruN ./orte/tools/orte-ps/Makefile.am ../openmpi-1.4.2.new/orte/tools/orte-ps/Makefile.am
--- ./orte/tools/orte-ps/Makefile.am	2009-12-08 13:36:41.000000000 -0700
+++ ../openmpi-1.4.2.new/orte/tools/orte-ps/Makefile.am	2010-05-30 08:47:26.000000000 -0600
@@ -40,7 +40,7 @@
 endif # OMPI_INSTALL_BINARIES
 
 orte_ps_SOURCES = orte-ps.c
-orte_ps_LDADD = $(top_builddir)/orte/libopen-rte.la
+orte_ps_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 
 endif # !ORTE_DISABLE_FULL_SUPPORT
 
diff -ruN ./orte/tools/orte-ps/Makefile.in ../openmpi-1.4.2.new/orte/tools/orte-ps/Makefile.in
--- ./orte/tools/orte-ps/Makefile.in	2010-05-04 10:08:02.000000000 -0600
+++ ../openmpi-1.4.2.new/orte/tools/orte-ps/Makefile.in	2010-05-30 08:47:38.000000000 -0600
@@ -255,7 +255,10 @@
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@am_orte_ps_OBJECTS =  \
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@	orte-ps.$(OBJEXT)
 orte_ps_OBJECTS = $(am_orte_ps_OBJECTS)
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_ps_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la
+am__DEPENDENCIES_1 =
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_ps_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1) \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1)
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)/opal/include -I$(top_builddir)/orte/include -I$(top_builddir)/ompi/include -I$(top_builddir)/opal/mca/paffinity/linux/plpa/src/libplpa
 depcomp = $(SHELL) $(top_srcdir)/config/depcomp
 am__depfiles_maybe = depfiles
@@ -777,6 +780,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -961,7 +965,7 @@
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@nodist_man_MANS = $(man_pages)
 @OMPI_INSTALL_BINARIES_TRUE@@ORTE_DISABLE_FULL_SUPPORT_FALSE@dist_pkgdata_DATA = help-orte-ps.txt
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_ps_SOURCES = orte-ps.c
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_ps_LDADD = $(top_builddir)/orte/libopen-rte.la
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orte_ps_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:
diff -ruN ./orte/tools/orterun/Makefile.am ../openmpi-1.4.2.new/orte/tools/orterun/Makefile.am
--- ./orte/tools/orterun/Makefile.am	2009-12-08 13:36:44.000000000 -0700
+++ ../openmpi-1.4.2.new/orte/tools/orterun/Makefile.am	2010-05-30 08:48:54.000000000 -0600
@@ -55,7 +55,7 @@
         debuggers.h \
         debuggers.c
  
-orterun_LDADD = $(top_builddir)/orte/libopen-rte.la
+orterun_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 
 endif # !ORTE_DISABLE_FULL_SUPPORT
 
diff -ruN ./orte/tools/orterun/Makefile.in ../openmpi-1.4.2.new/orte/tools/orterun/Makefile.in
--- ./orte/tools/orterun/Makefile.in	2010-05-04 10:08:04.000000000 -0600
+++ ../openmpi-1.4.2.new/orte/tools/orterun/Makefile.in	2010-05-30 08:49:44.000000000 -0600
@@ -264,7 +264,10 @@
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@	orterun.$(OBJEXT) \
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@	debuggers.$(OBJEXT)
 orterun_OBJECTS = $(am_orterun_OBJECTS)
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orterun_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la
+am__DEPENDENCIES_1 =
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orterun_DEPENDENCIES = $(top_builddir)/orte/libopen-rte.la \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1) \
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@	$(am__DEPENDENCIES_1)
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)/opal/include -I$(top_builddir)/orte/include -I$(top_builddir)/ompi/include -I$(top_builddir)/opal/mca/paffinity/linux/plpa/src/libplpa
 depcomp = $(SHELL) $(top_srcdir)/config/depcomp
 am__depfiles_maybe = depfiles
@@ -786,6 +789,7 @@
 PACKAGE_NAME = @PACKAGE_NAME@
 PACKAGE_STRING = @PACKAGE_STRING@
 PACKAGE_TARNAME = @PACKAGE_TARNAME@
+PACKAGE_URL = @PACKAGE_URL@
 PACKAGE_VERSION = @PACKAGE_VERSION@
 PATH_SEPARATOR = @PATH_SEPARATOR@
 PERL = @PERL@
@@ -976,7 +980,7 @@
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@        debuggers.h \
 @ORTE_DISABLE_FULL_SUPPORT_FALSE@        debuggers.c
 
-@ORTE_DISABLE_FULL_SUPPORT_FALSE@orterun_LDADD = $(top_builddir)/orte/libopen-rte.la
+@ORTE_DISABLE_FULL_SUPPORT_FALSE@orterun_LDADD = $(top_builddir)/orte/libopen-rte.la $(ORTE_WRAPPER_EXTRA_LDFLAGS) $(ORTE_WRAPPER_EXTRA_LIBS)
 all: all-am
 
 .SUFFIXES:

