diff --git a/config/ompi_config_pthreads.m4 b/config/ompi_config_pthreads.m4
index 3f073d8..1ff8ada 100644
--- a/config/ompi_config_pthreads.m4
+++ b/config/ompi_config_pthreads.m4
@@ -267,6 +267,14 @@ AC_DEFUN([OMPI_INTL_POSIX_THREADS_SPECIAL_FLAGS_C], [
 #
 # C compiler
 #
+case "${host_cpu}-${host_os}" in
+  *solaris*)
+    pflags="-pthread -pthreads -mt"
+  ;;
+  *)
+    pflags="-Kthread -kthread -pthread -pthreads -mt -mthreads"
+  ;;
+esac
 if test "$ompi_pthread_c_success" = "0"; then
   for pf in $pflags; do
     AC_MSG_CHECKING([if C compiler and POSIX threads work with $pf])
@@ -294,6 +302,14 @@ AC_DEFUN([OMPI_INTL_POSIX_THREADS_SPECIAL_FLAGS_CXX], [
 # C++ compiler
 #
 if test "$ompi_pthread_cxx_success" = "0"; then
+  case "${host_cpu}-${host_os}" in
+    *solaris*)
+      pflags="-pthread -pthreads -mt"
+    ;;
+    *)
+      pflags="-Kthread -kthread -pthread -pthreads -mt -mthreads"
+    ;;
+  esac
   for pf in $pflags; do
     AC_MSG_CHECKING([if C++ compiler and POSIX threads work with $pf])
     CXXFLAGS="$orig_CXXFLAGS $pf"
@@ -320,6 +336,14 @@ AC_DEFUN([OMPI_INTL_POSIX_THREADS_SPECIAL_FLAGS_FC], [
 # Fortran compiler
 #
 if test "$ompi_pthread_fortran_success" = "0" -a "$OMPI_WANT_FORTRAN_BINDINGS" = "1" -a $ompi_fortran_happy -eq 1; then
+  case "${host_cpu}-${host_os}" in
+    *solaris*)
+      pflags="-pthread -pthreads -mt"
+    ;;
+    *)
+      pflags="-Kthread -kthread -pthread -pthreads -mt -mthreads"
+    ;;
+  esac
   for pf in $pflags; do
     AC_MSG_CHECKING([if Fortran compiler and POSIX threads work with $pf])
     FCFLAGS="$orig_FCFLAGS $pf"
