Index: src/topology.c
===================================================================
--- src/topology.c	(révision 3750)
+++ src/topology.c	(copie de travail)
@@ -1856,6 +1856,8 @@
   /*
    * Now that objects are numbered, take distance matrices from backends and put them in the main topology
    */
+  hwloc_restrict_distances(topology, HWLOC_RESTRICT_FLAG_ADAPT_DISTANCES);
+  hwloc_convert_distances_indexes_into_objects(topology);
   hwloc_finalize_logical_distances(topology);
 
 #  ifdef HWLOC_HAVE_XML

