Index: ompi/mpi/f77/file_read_at_f.c
===================================================================
--- ompi/mpi/f77/file_read_at_f.c	(revision 12683)
+++ ompi/mpi/f77/file_read_at_f.c	(working copy)
@@ -9,6 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
+ * Copyright (c) 2006      Cisco Systems, Inc.  All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -95,7 +96,7 @@
 					     c_type, c_status));
 
 #if OMPI_SIZEOF_FORTRAN_INTEGER != SIZEOF_INT
-   if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) &&& 
+   if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) &&
        MPI_STATUS_IGNORE != c_status) {
         MPI_Status_c2f(c_status, status);
     }
Index: ompi/mpi/f77/file_read_ordered_end_f.c
===================================================================
--- ompi/mpi/f77/file_read_ordered_end_f.c	(revision 12683)
+++ ompi/mpi/f77/file_read_ordered_end_f.c	(working copy)
@@ -9,6 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
+ * Copyright (c) 2006      Cisco Systems, Inc.  All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -89,7 +90,7 @@
     *ierr = OMPI_INT_2_FINT(MPI_File_read_ordered_end(c_fh, buf, c_status));
 
 #if OMPI_SIZEOF_FORTRAN_INTEGER != SIZEOF_INT
-    if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) &&& 
+    if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) && 
        MPI_STATUS_IGNORE != c_status) {
         MPI_Status_c2f(c_status, status);
     }
Index: ompi/mpi/f77/file_read_at_all_f.c
===================================================================
--- ompi/mpi/f77/file_read_at_all_f.c	(revision 12683)
+++ ompi/mpi/f77/file_read_at_all_f.c	(working copy)
@@ -9,6 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
+ * Copyright (c) 2006      Cisco Systems, Inc.  All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -95,7 +96,7 @@
                                                 c_status));
 
 #if OMPI_SIZEOF_FORTRAN_INTEGER != SIZEOF_INT
-   if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) &&& 
+   if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) &&
        MPI_STATUS_IGNORE != c_status) {
       MPI_Status_c2f(c_status, status);
    }
Index: ompi/mpi/f77/file_read_f.c
===================================================================
--- ompi/mpi/f77/file_read_f.c	(revision 12683)
+++ ompi/mpi/f77/file_read_f.c	(working copy)
@@ -9,6 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
+ * Copyright (c) 2006      Cisco Systems, Inc.  All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -90,7 +91,7 @@
                                          c_type, c_status));
 
 #if OMPI_SIZEOF_FORTRAN_INTEGER != SIZEOF_INT
-   if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) &&& 
+   if (MPI_SUCCESS == OMPI_FINT_2_INT(*ierr) && 
        MPI_STATUS_IGNORE != c_status) {
       MPI_Status_c2f(c_status, status);
    }

