--- orte/mca/oob/tcp/oob_tcp.c.orig	2007-12-18 11:21:08.000000000 -0800
+++ orte/mca/oob/tcp/oob_tcp.c	2007-12-18 11:22:29.000000000 -0800
@@ -1344,11 +1344,15 @@
         char name[32];
         opal_ifindextoname(i, name, sizeof(name));
         if (mca_oob_tcp_component.tcp_include != NULL &&
-            strstr(mca_oob_tcp_component.tcp_include,name) == NULL)
+            strstr(mca_oob_tcp_component.tcp_include,name) == NULL) {
+            opal_output(0, "TCP OOB skipping %s because it's not in include (%s)\n", name, mca_oob_tcp_component.tcp_include);
             continue;
+        }
         if (mca_oob_tcp_component.tcp_exclude != NULL &&
-            strstr(mca_oob_tcp_component.tcp_exclude,name) != NULL)
+            strstr(mca_oob_tcp_component.tcp_exclude,name) != NULL) {
+            opal_output(0, "TCP OOB skipping %s because it's in exclude (%s)\n", name, mca_oob_tcp_component.tcp_exclude);
             continue;
+        }
         opal_ifindextoaddr(i, (struct sockaddr*)&addr, sizeof(addr));
         if(opal_ifcount() > 1 && 
            opal_ifislocalhost((struct sockaddr*) &addr))
@@ -1356,6 +1360,7 @@
         if(ptr != contact_info) {
             ptr += sprintf(ptr, ";");
         }
+        opal_output(0, "TCP OOB adding interface: %s\n", name);
         ptr += sprintf(ptr, "tcp://%s:%d", inet_ntoa(addr.sin_addr),
                     ntohs(mca_oob_tcp_component.tcp_listen_port));
     }

