diff --git a/orte/mca/state/dvm/state_dvm.c b/orte/mca/state/dvm/state_dvm.c
index 0e7309c..5b1a841 100644
--- a/orte/mca/state/dvm/state_dvm.c
+++ b/orte/mca/state/dvm/state_dvm.c
@@ -267,6 +267,7 @@ void check_complete(int fd, short args, void *cbdata)
         if (jdata->state < ORTE_JOB_STATE_UNTERMINATED) {
             jdata->state = ORTE_JOB_STATE_TERMINATED;
         }
+        opal_output(0, "%s JOB %s HAS TERMINATED", ORTE_NAME_PRINT(ORTE_PROC_MY_NAME), ORTE_JOBID_PRINT(jdata->jobid));
     }
 
     /* tell the IOF that the job is complete */
diff --git a/orte/tools/orte-dvm/orte-dvm.c b/orte/tools/orte-dvm/orte-dvm.c
index 3cdf585..f9a969a 100644
--- a/orte/tools/orte-dvm/orte-dvm.c
+++ b/orte/tools/orte-dvm/orte-dvm.c
@@ -462,6 +462,11 @@ static void notify_requestor(int sd, short args, void *cbdata)
     int ret;
     opal_buffer_t *reply;
 
+opal_output(0, "%s NOTIFYING %s OF JOB %s COMPLETION",
+            ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),
+            ORTE_NAME_PRINT(&jdata->originator),
+            ORTE_JOBID_PRINT(jdata->jobid));
+
     /* notify the requestor */
     reply = OBJ_NEW(opal_buffer_t);
     /* see if there was any problem */
@@ -476,6 +481,7 @@ static void notify_requestor(int sd, short args, void *cbdata)
     /* we cannot cleanup the job object as we might
      * hit an error during transmission, so clean it
      * up in the send callback */
+
     OBJ_RELEASE(caddy);
 }
 
