Index: ompi/datatype/ompi_datatype_args.c
===================================================================
--- ompi/datatype/ompi_datatype_args.c	(revision 28778)
+++ ompi/datatype/ompi_datatype_args.c	(working copy)
@@ -129,7 +129,13 @@
         /* Recompute the data description packed size based on the optimization
          * for MPI_COMBINER_DUP.
          */
-        pArgs->total_pack_size = 2 * sizeof(int);
+        if (ompi_datatype_is_predefined(d[0])) {
+            /* MPI_COMBINER_DUP and d[0]->id */
+            pArgs->total_pack_size = 2 * sizeof(int);
+        } else {
+            /* don't pack the description of MPI_COMBINER_DUP itself */
+            pArgs->total_pack_size = 0;
+        }
         break;
 
     case MPI_COMBINER_CONTIGUOUS:
@@ -449,9 +455,10 @@
     }
     /* For duplicated datatype we don't have to store all the information */
     if( MPI_COMBINER_DUP == args->create_type ) {
-        position[0] = args->create_type;
-        position[1] = args->d[0]->id; /* On the OMPI - layer, copy the ompi_datatype.id */
-        return OMPI_SUCCESS;
+        ompi_datatype_t* temp_data = args->d[0];
+        return __ompi_datatype_pack_description(temp_data,
+                                                packed_buffer,
+                                                next_index );
     }
     position[0] = args->create_type;
     position[1] = args->ci;

