--- ../openmpi-1.8.1-gcc/ompi/mca/btl/openib/btl_openib.c	2014-04-22 16:51:26.000000000 +0200
+++ ./ompi/mca/btl/openib/btl_openib.c	2014-05-27 23:49:45.690199577 +0200
@@ -917,7 +917,7 @@
     opal_bitmap_t* reachable)
 {
     mca_btl_openib_module_t* openib_btl = (mca_btl_openib_module_t*)btl;
-    int i,j, rc, local_procs;
+    int i,j, rc, local_procs = 0;
     int rem_subnet_id_port_cnt;
     int lcl_subnet_id_port_cnt = 0;
     int btl_rank = 0;
@@ -952,7 +952,7 @@
         return rc;
     }
 
-    for (i = 0, local_procs = 0 ; i < (int) nprocs; i++) {
+    for (i = 0 ; i < (int) nprocs; i++) {
         struct ompi_proc_t* ompi_proc = ompi_procs[i];
         mca_btl_openib_proc_t* ib_proc;
         int remote_matching_port;

