Index: opal/tools/wrappers/Makefile.am
===================================================================
--- opal/tools/wrappers/Makefile.am	(revision 28924)
+++ opal/tools/wrappers/Makefile.am	(working copy)
@@ -47,8 +47,8 @@
 pkgconfig_DATA = opal.pc
 
 install-exec-hook:
-	(cd $(DESTDIR)$(bindir); rm -f opalcc$(EXEEXT); $(LN_S) opal_wrapper opalcc)
-	(cd $(DESTDIR)$(bindir); rm -f opalc++$(EXEEXT); $(LN_S) opal_wrapper opalc++)
+	(cd $(DESTDIR)$(bindir); rm -f opalcc$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) opalcc$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f opalc++$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) opalc++$(EXEEXT))
 
 uninstall-local:
 	rm -f $(DESTDIR)$(bindir)/opalcc$(EXEEXT) \
Index: orte/tools/wrappers/Makefile.am
===================================================================
--- orte/tools/wrappers/Makefile.am	(revision 28924)
+++ orte/tools/wrappers/Makefile.am	(working copy)
@@ -33,8 +33,8 @@
 
 install-exec-hook-always:
 	test -z "$(bindir)" || $(MKDIR_P) "$(DESTDIR)$(bindir)"
-	(cd $(DESTDIR)$(bindir); rm -f ortecc$(EXEEXT); $(LN_S) opal_wrapper ortecc)
-	(cd $(DESTDIR)$(bindir); rm -f ortec++$(EXEEXT); $(LN_S) opal_wrapper ortec++)
+	(cd $(DESTDIR)$(bindir); rm -f ortecc$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) ortecc$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f ortec++$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) ortec++$(EXEEXT))
 
 uninstall-local-always:
 	rm -f $(DESTDIR)$(bindir)/ortecc$(EXEEXT) \
@@ -42,7 +42,7 @@
 
 if CASE_SENSITIVE_FS
 install-exec-hook: install-exec-hook-always
-	(cd $(DESTDIR)$(bindir); rm -f orteCC$(EXEEXT); $(LN_S) opal_wrapper orteCC)
+	(cd $(DESTDIR)$(bindir); rm -f orteCC$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) orteCC$(EXEEXT))
 
 install-data-hook:
 	(cd $(DESTDIR)$(pkgdatadir); rm -f orteCC-wrapper-data.txt; $(LN_S) ortec++-wrapper-data.txt orteCC-wrapper-data.txt)
Index: ompi/tools/wrappers/Makefile.am
===================================================================
--- ompi/tools/wrappers/Makefile.am	(revision 28924)
+++ ompi/tools/wrappers/Makefile.am	(working copy)
@@ -75,13 +75,13 @@
 install-exec-hook-always:
 	test -z "$(bindir)" || $(MKDIR_P) "$(DESTDIR)$(bindir)"
 	(cp mpijavac.pl $(DESTDIR)$(bindir))
-	(cd $(DESTDIR)$(bindir); rm -f mpicc$(EXEEXT); $(LN_S) opal_wrapper mpicc)
-	(cd $(DESTDIR)$(bindir); rm -f mpic++$(EXEEXT); $(LN_S) opal_wrapper mpic++)
-	(cd $(DESTDIR)$(bindir); rm -f mpicxx$(EXEEXT); $(LN_S) opal_wrapper mpicxx)
-	(cd $(DESTDIR)$(bindir); rm -f mpifort$(EXEEXT); $(LN_S) opal_wrapper mpifort)
-	(cd $(DESTDIR)$(bindir); rm -f mpif77$(EXEEXT); $(LN_S) mpifort mpif77)
-	(cd $(DESTDIR)$(bindir); rm -f mpif90$(EXEEXT); $(LN_S) mpifort mpif90)
-	(cd $(DESTDIR)$(bindir); chmod +x mpijavac.pl; rm -f mpijavac; $(LN_S) mpijavac.pl mpijavac)
+	(cd $(DESTDIR)$(bindir); rm -f mpicc$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) mpicc$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f mpic++$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) mpic++$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f mpicxx$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) mpicxx$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f mpifort$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) mpifort$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f mpif77$(EXEEXT); $(LN_S) mpifort$(EXEEXT) mpif77$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); rm -f mpif90$(EXEEXT); $(LN_S) mpifort$(EXEEXT) mpif90$(EXEEXT))
+	(cd $(DESTDIR)$(bindir); chmod +x mpijavac.pl; rm -f mpijavac$(EXEEXT); $(LN_S) mpijavac.pl mpijavac$(EXEEXT))
 
 install-data-hook-always:
 	(cd $(DESTDIR)$(pkgdatadir); rm -f mpicxx-wrapper-data.txt; $(LN_S) mpic++-wrapper-data.txt mpicxx-wrapper-data.txt)
@@ -103,12 +103,12 @@
 	$(DESTDIR)$(pkgconfigdir)/ompi-f77.pc \
 	$(DESTDIR)$(pkgconfigdir)/ompi-f90.pc \
 	$(DESTDIR)$(pkgconfigdir)/ompi-fort.pc \
-	$(DESTDIR)$(bindir)/mpijavac \
+	$(DESTDIR)$(bindir)/mpijavac$(EXEEXT) \
 	$(DESTDIR)$(bindir)/mpijavac.pl
 
 if CASE_SENSITIVE_FS
 install-exec-hook: install-exec-hook-always
-	(cd $(DESTDIR)$(bindir); rm -f mpiCC$(EXEEXT); $(LN_S) opal_wrapper mpiCC)
+	(cd $(DESTDIR)$(bindir); rm -f mpiCC$(EXEEXT); $(LN_S) opal_wrapper$(EXEEXT) mpiCC$(EXEEXT))
 
 install-data-hook: install-data-hook-always
 	(cd $(DESTDIR)$(pkgdatadir); rm -f mpiCC-wrapper-data.txt; $(LN_S) mpic++-wrapper-data.txt mpiCC-wrapper-data.txt)
