diff -r -U 5 openmpi-1.3.2/opal/util/if.c openmpi-1.3.2.edited/opal/util/if.c
--- openmpi-1.3.2/opal/util/if.c	2009-04-16 20:02:42.000000000 +0100
+++ openmpi-1.3.2.edited/opal/util/if.c	2009-04-23 16:18:09.000000000 +0100
@@ -258,11 +258,12 @@
         struct ifreq* ifr = (struct ifreq*) ptr;
         opal_if_t intf;
         opal_if_t *intf_ptr;
         int length;
 
-        OMPI_DEBUG_ZERO(intf);
+        /* Again, make valgrind and purify happy - this isn't performance critical. */
+        memset(&intf, 0, sizeof(intf));
         OBJ_CONSTRUCT(&intf, opal_list_item_t);
 
         /* compute offset for entries */
 #ifdef HAVE_STRUCT_SOCKADDR_SA_LEN
         length = sizeof(struct sockaddr);

