Index: autogen.sh
===================================================================
--- autogen.sh	(revision 7061)
+++ autogen.sh	(working copy)
@@ -22,18 +22,29 @@
 ##############################################################################
 #
 # User-definable parameters (search path and minimum supported versions)
-#
+# 
+#   Note: use ';' to separate parameters
 ##############################################################################
 
-ompi_aclocal_search="aclocal"
-ompi_autoheader_search="autoheader"
-ompi_autoconf_search="autoconf"
-ompi_libtoolize_search="libtoolize glibtoolize"
-ompi_automake_search="automake"
+if [ -z "$ACLOCAL" ]; then	ompi_aclocal_search="aclocal"
+else ompi_aclocal_search=$ACLOCAL
+fi
+if [ -z "$AUTOHEADER" ]; then	ompi_autoheader_search="autoheader"
+else ompi_autoheader_search=$AUTOHEADER
+fi
+if [ -z "$AUTOCONF" ]; then	ompi_autoconf_search="autoconf"
+else ompi_autoconf_search=$AUTOCONF
+fi
+if [ -z "$LIBTOOLIZE" ]; then	ompi_libtoolize_search="libtoolize;glibtoolize"
+else ompi_libtoolize_search=$LIBTOOLIZE
+fi
+if [ -z "$AUTOMAKE" ]; then	ompi_automake_search="automake"
+else ompi_automake_search=$AUTOMAKE
+fi
 
 ompi_automake_version="1.7"
 ompi_autoconf_version="2.58"
-ompi_libtool_version="1.5"
+ompi_libtool_version="1.5.18"
 
 
 ##############################################################################
@@ -46,13 +57,6 @@
 ompi_autoheader_version="$ompi_autoconf_version"
 ompi_libtoolize_version="$ompi_libtool_version"
 
-# program names to execute
-ompi_aclocal=""
-ompi_autoheader=""
-ompi_autoconf=""
-ompi_libtoolize=""
-ompi_automake=""
-
 mca_no_configure_components_file="config/mca_no_configure_components.m4"
 mca_no_config_list_file="mca_no_config_list"
 mca_no_config_env_file="mca_no_config_env"
@@ -137,11 +141,14 @@
     local version="0.0.0"
     local min_version="99.99.99"
     local found=0
+    local tmpIFS=$IFS
 
+    IFS=";"
     eval "min_version=\"\$ompi_${app_name}_version\""
 
     eval "search_path=\"\$ompi_${app_name}_search\""
     for i in $search_path ; do
+        IFS=$tmpIFS
         version="`${i} --version 2>&1`"
         if test "$?" != 0 ; then
             continue
@@ -157,6 +164,8 @@
         fi
     done
 
+    IFS=$tmpIFS
+
     if test "$found" = "0" ; then
 	cat <<EOF
 I could not find a recent enough copy of ${app_name}.
@@ -1057,11 +1066,13 @@
 
 # find all the apps we are going to run
 find_app "aclocal"
+echo "ompi_aclocal: $ompi_aclocal"
 find_app "autoheader"
 find_app "autoconf"
 find_app "libtoolize"
 find_app "automake"
 
+
 # do the work
 if test "$want_local" = "yes"; then
     process_dir . $ompidir

