--- origsrc/openmpi-1.8.5rc2/ompi/mpi/fortran/use-mpi-f08/Makefile.am	2015-04-05 20:40:24.000000000 +0200
+++ src/openmpi-1.8.5rc2/ompi/mpi/fortran/use-mpi-f08/Makefile.am	2015-04-22 15:39:46.739793600 +0200
@@ -805,6 +805,7 @@ endif
 libmpi_usempif08_la_LIBADD = \
         $(module_sentinel_file) \
         $(OMPI_MPIEXT_USEMPIF08_LIBS) \
+        $(top_builddir)/ompi/mpi/fortran/mpif-h/libmpi_mpifh.la \
         $(top_builddir)/ompi/libmpi.la
 libmpi_usempif08_la_DEPENDENCIES = $(module_sentinel_file)
 libmpi_usempif08_la_LDFLAGS = -version-info $(libmpi_usempif08_so_version)

