diff --git a/ompi/mca/io/romio/romio/adio/common/shfp_fname.c b/ompi/mca/io/romio/romio/adio/common/shfp_fname.c
index 024ced5..98ec101 100644
--- a/ompi/mca/io/romio/romio/adio/common/shfp_fname.c
+++ b/ompi/mca/io/romio/romio/adio/common/shfp_fname.c
@@ -2,6 +2,8 @@
 /* 
  *
  *   Copyright (C) 1997 University of Chicago. 
+ *   Copyright (c) 2014 Research Organization for Information Science
+ *                      and Technology (RIST). All rights reserved.
  *   See COPYRIGHT notice in top-level directory.
  */
 
@@ -51,7 +53,7 @@ void ADIOI_Shfp_fname(ADIO_File fd, int rank)
 	    slash = strrchr(fd->shared_fp_fname, '/');
 #endif
 	    ADIOI_Strncpy(slash + 1, ".", 2);
-	    len = 256 - (slash+2 - fd->shared_fp_fname);
+	    len = 255 - (slash+2 - fd->shared_fp_fname);
 	    ADIOI_Strncpy(slash + 2, ptr + 1, len);
 	}
 	    

