diff --git a/opal/mca/base/mca_base_var.c b/opal/mca/base/mca_base_var.c
index 7b55eb8..c043c06 100644
--- a/opal/mca/base/mca_base_var.c
+++ b/opal/mca/base/mca_base_var.c
@@ -1682,7 +1682,11 @@ static int var_value_string (mca_base_var_t *var, char **value_string)
 
         ret = (0 > ret) ? OPAL_ERR_OUT_OF_RESOURCE : OPAL_SUCCESS;
     } else {
-        ret = var->mbv_enumerator->string_from_value(var->mbv_enumerator, value->intval, &tmp);
+        if (MCA_BASE_VAR_TYPE_BOOL == var->mbv_type) {
+            ret = var->mbv_enumerator->string_from_value(var->mbv_enumerator, value->boolval, &tmp);
+        } else {
+            ret = var->mbv_enumerator->string_from_value(var->mbv_enumerator, value->intval, &tmp);
+        }
 
         *value_string = strdup (tmp);
         if (NULL == value_string) {

