diff --git a/opal/mca/btl/openib/btl_openib_proc.c b/opal/mca/btl/openib/btl_openib_proc.c
index 2d622fe..6e17320 100644
--- a/opal/mca/btl/openib/btl_openib_proc.c
+++ b/opal/mca/btl/openib/btl_openib_proc.c
@@ -12,6 +12,8 @@
  * Copyright (c) 2007-2008 Cisco Systems, Inc.  All rights reserved.
  * Copyright (c) 2006-2007 Voltaire All rights reserved.
  * Copyright (c) 2014      Intel, Inc. All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  *
  * Additional copyrights may follow
@@ -159,6 +161,13 @@ mca_btl_openib_proc_t* mca_btl_openib_proc_create(opal_proc_t* proc)
     if (0 == msg_size) {
         return NULL;
     }
+    /* do NOT use ib as a loopback interface */
+    if (NULL != proc->proc_hostname) {
+        char * h = opal_proc_local_get()->proc_hostname;
+        if (strcmp(h, proc->proc_hostname) == 0) {
+            return NULL;
+        }
+    }
 
     /* Message was packed in btl_openib_component.c; the format is
        listed in a comment in that file */

