Index: ompi/mpi/cxx/mpicxx.cc
===================================================================
--- ompi/mpi/cxx/mpicxx.cc	(revision 8871)
+++ ompi/mpi/cxx/mpicxx.cc	(working copy)
@@ -31,7 +31,8 @@
 
 // return  codes
 const int SUCCESS = MPI_SUCCESS;
-const int ERR_BUFFER = MPI_ERR_COUNT;
+const int ERR_BUFFER = MPI_ERR_BUFFER;
+const int ERR_COUNT = MPI_ERR_COUNT;
 const int ERR_TYPE = MPI_ERR_TYPE;
 const int ERR_TAG = MPI_ERR_TAG;
 const int ERR_COMM = MPI_ERR_COMM;
@@ -91,6 +92,7 @@
 const Datatype SHORT(MPI_SHORT);
 const Datatype INT(MPI_INT);
 const Datatype LONG(MPI_LONG);
+const Datatype SIGNED_CHAR(MPI_SIGNED_CHAR);
 const Datatype UNSIGNED_CHAR(MPI_UNSIGNED_CHAR);
 const Datatype UNSIGNED_SHORT(MPI_UNSIGNED_SHORT);
 const Datatype UNSIGNED(MPI_UNSIGNED);
Index: ompi/mpi/cxx/constants.h
===================================================================
--- ompi/mpi/cxx/constants.h	(revision 8871)
+++ ompi/mpi/cxx/constants.h	(working copy)
@@ -200,7 +200,9 @@
 
 // null handles
 OMPI_DECLSPEC extern const Group        GROUP_NULL;
+#if OMPI_WANT_MPI2_ONE_SIDED
 OMPI_DECLSPEC extern const Win          WIN_NULL;
+#endif
 OMPI_DECLSPEC extern const Info         INFO_NULL;
 //OMPI_DECLSPEC extern const Comm         COMM_NULL;
 //OMPI_DECLSPEC extern const MPI_Comm     COMM_NULL;
Index: NEWS
===================================================================
--- NEWS	(revision 8871)
+++ NEWS	(working copy)
@@ -29,6 +29,7 @@
 1.0.2
 -----
 
+- Fixed missing constants in MPI C++ bindings.
 - Fixed some errors caused by threading issues.
 - Fixed openib BTL flow control logic to not overrun the number of
   send wqes available.

