--- config/ltmain.sh.orig	2011-06-09 12:50:08.911201988 +0200
+++ config/ltmain.sh	2011-06-09 12:51:20.530015482 +0200
@@ -5099,7 +5099,7 @@
 	      # Search the libtool library
 	      lib="$searchdir/lib${name}${search_ext}"
 	      if test -f "$lib"; then
-		if test "$search_ext" = ".la"; then
+		if test "$search_ext" = ".la" -a "$lib" != "`pwd`/$outputname"; then
 		  found=yes
 		else
 		  found=no

