Index: ompi/proc/proc.c
===================================================================
--- ompi/proc/proc.c	(revision 17898)
+++ ompi/proc/proc.c	(working copy)
@@ -192,6 +192,11 @@
         size_t datalen;
         orte_vpid_t nodeid;
 
+        /* Don't reset the information determined about the current
+           process during the init step.  Saves time and problems if
+           modex is unimplemented */
+        if (ompi_proc_local() == proc) continue;
+
         if (OPAL_EQUAL != orte_util_compare_name_fields(ORTE_NS_CMP_JOBID,
                                                  &ompi_proc_local_proc->proc_name,
                                                  &proc->proc_name)) {
