Index: config/ompi_config_asm.m4
===================================================================
--- config/ompi_config_asm.m4	(revision 7292)
+++ config/ompi_config_asm.m4	(working copy)
@@ -752,7 +752,7 @@
         OMPI_GCC_INLINE_ASSIGN='"bis zero,zero,%0" : "=&r"(ret)'
     ;;
 
-    mips-*)
+    mips-*|mpis64-*)
         # Should really find some way to make sure that we are on
         # a MIPS III machine (r4000 and later)
         ompi_cv_asm_arch="MIPS"

