diff -r f567823ec31b opal/mca/base/mca_base_component_find.c
--- a/opal/mca/base/mca_base_component_find.c	Fri Jul 31 17:11:48 2009 +0200
+++ b/opal/mca/base/mca_base_component_find.c	Mon Aug 03 14:22:01 2009 +0200
@@ -502,8 +502,8 @@
   component_struct = (mca_base_component_t*)lt_dlsym(component_handle, struct_name);
   if (NULL == component_struct) {
     opal_output_verbose(vl, 0, "mca: base: component_find: \"%s\" does not appear to be a valid "
-                        "%s MCA dynamic component (ignored)", 
-                        target_file->basename, target_file->type);
+                        "%s MCA dynamic component (ignored), lt_dlsym said: \"%s\"", 
+                        target_file->basename, target_file->type, lt_dlerror());
     free(mitem);
     free(struct_name);
     lt_dlclose(component_handle);

