diff -r 05b4993c3f72 ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in
--- a/ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in	Wed Apr 04 14:10:13 2012 -0700
+++ b/ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in	Fri Apr 06 20:07:47 2012 +0900
@@ -2765,8 +2765,8 @@
   @OMPI_FORTRAN_IGNORE_TKR_TYPE@, intent(in) :: sendbuf
   @OMPI_FORTRAN_IGNORE_TKR_PREDECL@ recvbuf
   @OMPI_FORTRAN_IGNORE_TKR_TYPE@ :: recvbuf
-  integer, intent(in) :: recvcounts
-  integer, dimension(*), intent(in) :: datatype
+  integer, dimension(*), intent(in) :: recvcounts
+  integer, intent(in) :: datatype
   integer, intent(in) :: op
   integer, intent(in) :: comm
   integer, intent(out) :: ierr
diff -r 05b4993c3f72 ompi/mpi/fortran/use-mpi-tkr/scripts/mpi-f90-interfaces.h.sh
--- a/ompi/mpi/fortran/use-mpi-tkr/scripts/mpi-f90-interfaces.h.sh	Wed Apr 04 14:10:13 2012 -0700
+++ b/ompi/mpi/fortran/use-mpi-tkr/scripts/mpi-f90-interfaces.h.sh	Fri Apr 06 20:07:47 2012 +0900
@@ -2365,7 +2365,7 @@
 
 subroutine ${proc}(fh, buf, count, datatype, request&
         , ierr)
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -2479,7 +2479,7 @@
 
 subroutine ${proc}(fh, buf, count, datatype, request&
         , ierr)
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -2757,7 +2757,7 @@
 subroutine ${proc}(fh, buf, count, datatype, status&
         , ierr)
   include 'mpif-config.h'
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -2814,7 +2814,7 @@
 subroutine ${proc}(fh, buf, count, datatype, status&
         , ierr)
   include 'mpif-config.h'
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -2869,7 +2869,7 @@
     cat <<EOF
 
 subroutine ${proc}(fh, buf, count, datatype, ierr)
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -2924,7 +2924,7 @@
 
 subroutine ${proc}(fh, buf, status, ierr)
   include 'mpif-config.h'
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, dimension(MPI_STATUS_SIZE), intent(out) :: status
   integer, intent(out) :: ierr
@@ -3206,7 +3206,7 @@
 subroutine ${proc}(fh, buf, count, datatype, status&
         , ierr)
   include 'mpif-config.h'
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -3261,7 +3261,7 @@
     cat <<EOF
 
 subroutine ${proc}(fh, buf, count, datatype, ierr)
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -3316,7 +3316,7 @@
 
 subroutine ${proc}(fh, buf, status, ierr)
   include 'mpif-config.h'
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, dimension(MPI_STATUS_SIZE), intent(out) :: status
   integer, intent(out) :: ierr
@@ -3371,7 +3371,7 @@
 subroutine ${proc}(fh, buf, count, datatype, status&
         , ierr)
   include 'mpif-config.h'
-  integer, intent(inout) :: fh
+  integer, intent(in) :: fh
   ${type} :: buf
   integer, intent(in) :: count
   integer, intent(in) :: datatype
@@ -6001,7 +6001,7 @@
   integer, intent(in) :: incount
   integer, intent(in) :: datatype
   ${type} :: outbuf
-  integer, intent(out) :: outsize
+  integer, intent(in) :: outsize
   integer, intent(inout) :: position
   integer, intent(in) :: comm
   integer, intent(out) :: ierr
