Index: ompi/mca/coll/ml/coll_ml_inlines.h
===================================================================
--- ompi/mca/coll/ml/coll_ml_inlines.h	(revision 32658)
+++ ompi/mca/coll/ml/coll_ml_inlines.h	(working copy)
@@ -192,7 +192,7 @@
                 !out_of_resource) {
                 */
         if (((&coll_op->full_message != coll_op->fragment_data.message_descriptor) &&
-	     !out_of_resource) || IS_COLL_SYNCMEM(coll_op)) {
+	     !out_of_resource)) {
             /* non-zero offset ==> this is not fragment 0 */
             CHECK_AND_RECYCLE(coll_op);
         }

