Index: oshmem/shmem/fortran/shmem_int4_cswap_f.c
===================================================================
--- oshmem/shmem/fortran/shmem_int4_cswap_f.c	(revision 32657)
+++ oshmem/shmem/fortran/shmem_int4_cswap_f.c	(working copy)
@@ -2,6 +2,8 @@
  * Copyright (c) 2013      Mellanox Technologies, Inc.
  *                         All rights reserved.
  * Copyright (c) 2013 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -38,7 +40,7 @@
 
     MCA_ATOMIC_CALL(cswap(FPTR_2_VOID_PTR(target), 
         (void *)&out_value, 
-        (const void*)(&OMPI_FINT_2_INT(*cond)), 
+        (const void*)(OMPI_PFINT_2_PINT(cond)), 
         FPTR_2_VOID_PTR(value), 
         sizeof(out_value), 
         OMPI_FINT_2_INT(*pe)));
Index: oshmem/shmem/fortran/shmem_int8_cswap_f.c
===================================================================
--- oshmem/shmem/fortran/shmem_int8_cswap_f.c	(revision 32657)
+++ oshmem/shmem/fortran/shmem_int8_cswap_f.c	(working copy)
@@ -2,6 +2,8 @@
  * Copyright (c) 2013      Mellanox Technologies, Inc.
  *                         All rights reserved.
  * Copyright (c) 2013 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -38,7 +40,7 @@
 
     MCA_ATOMIC_CALL(cswap(FPTR_2_VOID_PTR(target), 
         (void *)&out_value, 
-        (const void*)(&OMPI_FINT_2_INT(*cond)), 
+        (const void*)(OMPI_PFINT_2_PINT(cond)), 
         FPTR_2_VOID_PTR(value), 
         sizeof(out_value), 
         OMPI_FINT_2_INT(*pe)));
Index: ompi/mpi/fortran/base/fint_2_int.h
===================================================================
--- ompi/mpi/fortran/base/fint_2_int.h	(revision 32657)
+++ ompi/mpi/fortran/base/fint_2_int.h	(working copy)
@@ -11,6 +11,8 @@
  *                         All rights reserved.
  * Copyright (c) 2011-2012 Cisco Systems, Inc.  All rights reserved.
  * Copyright (c) 2012      Oracle and/or its affiliates.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -37,6 +39,7 @@
   #define OMPI_SINGLE_NAME_CONVERT(a) a
   #define OMPI_INT_2_FINT(a) a
   #define OMPI_FINT_2_INT(a) a
+  #define OMPI_PFINT_2_PINT(a) a
   #define OMPI_ARRAY_FINT_2_INT_ALLOC(in, n) 
   #define OMPI_ARRAY_FINT_2_INT(in, n)
   #define OMPI_2_DIM_ARRAY_FINT_2_INT(in, n, dim2) 
@@ -53,6 +56,7 @@
   #define OMPI_SINGLE_NAME_CONVERT(a) &c_##a
   #define OMPI_INT_2_FINT(a) a
   #define OMPI_FINT_2_INT(a) (int) (a)
+  #define OMPI_PFINT_2_PINT(a) (int *) (a)
 
   /* This is for OUT parameters. Does only alloc */
   #define OMPI_ARRAY_FINT_2_INT_ALLOC(in, n) \
@@ -109,6 +113,7 @@
   #define OMPI_SINGLE_NAME_CONVERT(a) &c_##a
   #define OMPI_INT_2_FINT(a) (MPI_Fint)(a)
   #define OMPI_FINT_2_INT(a) (a)
+  #define OMPI_PFINT_2_PINT(a) a
 
   /* This is for OUT parameters. Does only alloc */
   #define OMPI_ARRAY_FINT_2_INT_ALLOC(in, n) \

