--- ompi/proc/proc.c.orig	2013-09-30 17:03:59.000000000 -0400
+++ ompi/proc/proc.c	2013-09-30 17:12:09.000000000 -0400
@@ -677,8 +677,8 @@
 
             if (full_info) {
                 int32_t num_recvd_entries;
-                orte_std_cntr_t cnt;
-                orte_std_cntr_t j;
+                int32_t cnt;
+                int32_t j;
 
                 /* unpack the number of entries for this proc */
                 cnt = 1;
@@ -721,7 +721,7 @@
 #endif
                 if (ompi_process_info.num_procs < ompi_hostname_cutoff) {
                     /* retrieve the hostname */
-                    rc = opal_db.fetch_pointer((opal_identifier_t*)&new_name, ORTE_DB_HOSTNAME,
+                    rc = opal_db.fetch_pointer((opal_identifier_t*)&new_name, OMPI_DB_HOSTNAME,
                                                (void**)&new_hostname, OPAL_STRING);
                     if( OPAL_SUCCESS != rc ) {
                         new_hostname = NULL;
@@ -762,7 +762,7 @@
         } else {
             if (full_info) {
                 int32_t num_recvd_entries;
-                orte_std_cntr_t j, cnt;
+                int32_t j, cnt;
 
                 /* discard all keys: they are already locally known */
                 cnt = 1;
