--- orig/openmpi-1.10.2rc1/configure	2015-12-20 00:30:07.000000000 +0900
+++ openmpi-1.10.2rc1/configure	2015-12-22 09:25:27.837776888 +0900
@@ -58597,7 +58597,7 @@
     env | grep ompi_scope
     ompi_scope_index=`expr $ompi_scope_index + 1`
 
-    FCFLAGS="-I. $FCFLAGS"
+    FCFLAGS_save=$FCFLAGS; FCFLAGS="-I. $FCFLAGS"
 
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking if Fortran compiler supports USE...ONLY" >&5
 $as_echo_n "checking if Fortran compiler supports USE...ONLY... " >&6; }

