Index: opal/include/opal/sys/atomic.h
===================================================================
--- opal/include/opal/sys/atomic.h	(revision 27971)
+++ opal/include/opal/sys/atomic.h	(working copy)
@@ -389,7 +389,7 @@
    OPAL_HAVE_ATOMIC_CMPSET_32 is 1, then atomic_impl.h (below) will
    define a static inline version of it (in C, using
    atomic_cmpset_32()).  */
-#if OPAL_HAVE_INLINE_ATOMIC_ADD_32 || OPAL_HAVE_ATOMIC_CMPSET_32
+#if OPAL_HAVE_INLINE_ATOMIC_ADD_32 || OPAL_HAVE_INLINE_ATOMIC_CMPSET_32
 static inline
 #endif
 int32_t opal_atomic_add_32(volatile int32_t *addr, int delta);
@@ -399,7 +399,7 @@
    OPAL_HAVE_ATOMIC_CMPSET_32 is 1, then atomic_impl.h (below) will
    define a static inline version of it (in C, using
    atomic_cmpset_32()).  */
-#if OPAL_HAVE_INLINE_ATOMIC_SUB_32 || OPAL_HAVE_ATOMIC_CMPSET_32
+#if OPAL_HAVE_INLINE_ATOMIC_SUB_32 || OPAL_HAVE_INLINE_ATOMIC_CMPSET_32
 static inline
 #endif
 int32_t opal_atomic_sub_32(volatile int32_t *addr, int delta);
@@ -424,7 +424,7 @@
    OPAL_HAVE_ATOMIC_CMPSET_64 is 1, then atomic_impl.h (below) will
    define a static inline version of it (in C, using
    atomic_cmpset_64()).  */
-#if OPAL_HAVE_INLINE_ATOMIC_ADD_64 || OPAL_HAVE_ATOMIC_CMPSET_64
+#if OPAL_HAVE_INLINE_ATOMIC_ADD_64 || OPAL_HAVE_INLINE_ATOMIC_CMPSET_64
 static inline
 #endif
 int64_t opal_atomic_add_64(volatile int64_t *addr, int64_t delta);
@@ -434,7 +434,7 @@
    OPAL_HAVE_ATOMIC_CMPSET_64 is 1, then atomic_impl.h (below) will
    define a static inline version of it (in C, using
    atomic_cmpset_64()).  */
-#if OPAL_HAVE_INLINE_ATOMIC_SUB_64 || OPAL_HAVE_ATOMIC_CMPSET_64
+#if OPAL_HAVE_INLINE_ATOMIC_SUB_64 || OPAL_HAVE_INLINE_ATOMIC_CMPSET_64
 static inline
 #endif
 int64_t opal_atomic_sub_64(volatile int64_t *addr, int64_t delta);
