Index: opal/mca/hwloc/hwloc121/hwloc/src/traversal.c
===================================================================
--- opal/mca/hwloc/hwloc121/hwloc/src/traversal.c	(revision 25164)
+++ opal/mca/hwloc/hwloc121/hwloc/src/traversal.c	(working copy)
@@ -258,7 +258,7 @@
   switch (obj->type) {
   case HWLOC_OBJ_CACHE:
     if (verbose)
-      res = hwloc_snprintf(tmp, tmplen, "%s%lu%s%sline=%u",
+      res = hwloc_snprintf(tmp, tmplen, "%sCachesize=%lu%s%sCacheline=%u",
 			   prefix,
 			   (unsigned long) hwloc_memory_size_printf_value(obj->attr->cache.size, verbose),
 			   hwloc_memory_size_printf_unit(obj->attr->cache.size, verbose),

