--- tachyon/unix/Make-arch	2005-03-18 10:28:05.000000000 -0500
+++ tachyon-new/unix/Make-arch	2005-10-28 14:14:42.000000000 -0500
@@ -35,8 +35,10 @@
 	@echo "        irix6-64-thr - SGI IRIX 6.5.x   POSIX Threads, 64-bit "
 	@echo "   linux-beowulf-mpi - Scyld Linux MPI                        " 
 	@echo " linux-alpha-ccc-qsw - Linux Alpha, Compaq C, MPI, QSWnet     "
-	@echo "           linux-lam - Linux            MPI (OSC LAM)         "
-	@echo "        linux-lam-64 - Linux AMD64/EM64T, MPI, 64-bit         "
+	@echo "           linux-lam - Linux            MPI (LAM/MPI)         "
+	@echo "          linux-ompi - Linux            MPI (Open MPI)        "
+	@echo "        linux-lam-64 - Linux AMD64/EM64T, LAM/MPI, 64-bit     "
+	@echo "       linux-ompi-64 - Linux AMD64/EM64T, Open MPI, 64-bit    "
 	@echo "        linux-64-thr - Linux AMD64/EM64T, POSIX Threads, 64-bit"
 	@echo "           linux-thr - Linux            POSIX Threads         "
 	@echo "    linux-p4-icc-thr - Linux, Intel C compilers, P4-optimized  "
@@ -45,7 +47,8 @@
 	@echo "          macosx-thr - MacOS-X (aka Darwin), POSIX Threads    "
 	@echo "   solaris-ultra-hpc - Sun Solaris 8/9  ClusterTools 4.0 MPI  " 
 	@echo "         solaris-mpi - Sun Solaris 8/9  MPI (MPICH)           " 
-	@echo "         solaris-lam - Sun Solaris 8/9  MPI (OSC LAM)         " 
+	@echo "         solaris-lam - Sun Solaris 8/9  MPI (LAM/MPI)         " 
+	@echo "        solaris-ompi - Sun Solaris 8/9  MPI (Open MPI)        " 
 	@echo "         solaris-thr - Sun Solaris 8/9  UI/Sun Threads        " 
 	@echo "   solaris-ultra-thr - Sun Solaris 8/9  UI/Sun Threads, US-1/2" 
 	@echo "  solaris-ultra3-thr - Sun Solaris 8/9  UI/Sun Threads, US-III" 
@@ -83,7 +86,8 @@
 	@echo "solaris-ultra-thr-ogl - Sun Solaris 8/9 UI/Sun Threads, OpenGL"
 	@echo "solaris-ultra-hpc-ogl - Sun Solaris 8/9 Sun MPI, OpenGL       "
 	@echo "        linux-thr-ogl - Linux POSIX Threads, OpenGL           "
-	@echo "     linux-lam-64-ogl - Linux, 64-bit AMD64/EM64T, LAM MPI, OpenGL"
+	@echo "     linux-lam-64-ogl - Linux, 64-bit AMD64/EM64T, LAM/MPI, OpenGL"
+	@echo "    linux-ompi-64-ogl - Linux, 64-bit AMD64/EM64T, Open MPI, OpenGL"
 	@echo "linux-beowulf-mpi-ogl - Scyld Linux with MPI, OpenGL          "
 	@echo "  tru64-alpha-thr-ogl - Tru64/OSF1/DUnix POSIX Threads, OpenGL"
 	@echo ""
@@ -380,12 +384,22 @@
 solaris-lam:
 	$(MAKE) all \
 	"ARCH = solaris-lam" \
-	"CC = hcc" \
-	"CFLAGS = -v -xtransition -Xc -fast -xO5 -native -xdepend -fsimple=2 -I$(LAMHOME)/h -DSunOS $(MISCFLAGS) -DMPI" \
+	"CC = mpicc" \
+	"CFLAGS = -v -xtransition -Xc -fast -xO5 -native -xdepend -fsimple=2 -DSunOS $(MISCFLAGS) -DMPI" \
 	"AR = ar" \
 	"ARFLAGS = r" \
 	"STRIP = strip" \
-	"LIBS = -L. -L$(LAMHOME)/lib -ltachyon -lmpi $(MISCLIB) -lm -lsocket -lnsl"
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm -lsocket -lnsl"
+
+solaris-ompi:
+	$(MAKE) all \
+	"ARCH = solaris-ompi" \
+	"CC = mpicc" \
+	"CFLAGS = -v -xtransition -Xc -fast -xO5 -native -xdepend -fsimple=2 -DSunOS $(MISCFLAGS) -DMPI" \
+	"AR = ar" \
+	"ARFLAGS = r" \
+	"STRIP = strip" \
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm -lsocket -lnsl"
 
 solaris-ultra-hpc:
 	$(MAKE) all \
@@ -1077,53 +1091,101 @@
 	"RANLIB = ranlib" \
 	"LIBS = -L. -ltachyon -lmpi $(MISCLIB) -lm"
 
-# Linux x86 using LAM MPI
+# Linux x86 using LAM/MPI
 linux-lam:
 	$(MAKE) all \
 	"ARCH = linux-lam" \
-	"CC = hcc" \
-	"CFLAGS = -m486 -Wall -O6 -fomit-frame-pointer -ffast-math -I$(LAMHOME)/h -DLinux -DMPI $(MISCFLAGS)" \
+	"CC = mpicc" \
+	"CFLAGS = -m486 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI $(MISCFLAGS)" \
 	"AR = ar" \
 	"ARFLAGS = r" \
 	"STRIP = strip" \
 	"RANLIB = ranlib" \
-	"LIBS = -L. -L$(LAMHOME)/lib -ltachyon -lmpi $(MISCLIB) -lm"
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm"
 
-# Linux AMD64/EM64T using LAM MPI
+# Linux AMD64/EM64T using LAM/MPI
 linux-lam-64:
 	$(MAKE) all \
 	"ARCH = linux-lam-64" \
-	"CC = hcc" \
-	"CFLAGS = -m64 -Wall -O6 -fomit-frame-pointer -ffast-math -I$(LAMHOME)/h -DLinux -DMPI -DLP64 $(MISCFLAGS)" \
+	"CC = mpicc" \
+	"CFLAGS = -m64 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI -DLP64 $(MISCFLAGS)" \
 	"AR = ar" \
 	"ARFLAGS = r" \
 	"STRIP = strip" \
 	"RANLIB = ranlib" \
-	"LIBS = -L. -L$(LAMHOME)/lib -ltachyon -lmpi $(MISCLIB) -lm"
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm"
 
-# Linux AMD64/EM64T using LAM MPI, with OpenGL display
+# Linux AMD64/EM64T using LAM/MPI, with OpenGL display
 linux-lam-64-ogl:
 	$(MAKE) all \
 	"ARCH = linux-lam-64-ogl" \
-	"CC = hcc" \
-	"CFLAGS = -m64 -Wall -O6 -fomit-frame-pointer -ffast-math -I$(LAMHOME)/h -DLinux -DMPI -DLP64 -DUSEOPENGL $(LINUX_GLX_INCS) $(MISCFLAGS)" \
+	"CC = mpicc" \
+	"CFLAGS = -m64 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI -DLP64 -DUSEOPENGL $(LINUX_GLX_INCS) $(MISCFLAGS)" \
 	"AR = ar" \
 	"ARFLAGS = r" \
 	"STRIP = strip" \
 	"RANLIB = ranlib" \
-	"LIBS = -L. -L$(LAMHOME)/lib -ltachyon -lmpi $(LINUX_GLX_LIBS) $(MISCLIB) -lm"
+	"LIBS = -L. -ltachyon $(LINUX_GLX_LIBS) $(MISCLIB) -lm"
 
-# Linux x86 using LAM MPI and threads
+# Linux x86 using LAM/MPI and threads
 linux-lam-thr:
 	$(MAKE) all \
 	"ARCH = linux-lam-thr" \
-	"CC = hcc" \
-	"CFLAGS = -m486 -Wall -O6 -fomit-frame-pointer -ffast-math -I$(LAMHOME)/h -DLinux -DMPI -DTHR -D_REENTRANT $(MISCFLAGS)" \
+	"CC = mpicc" \
+	"CFLAGS = -m486 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI -DTHR -D_REENTRANT $(MISCFLAGS)" \
+	"AR = ar" \
+	"ARFLAGS = r" \
+	"STRIP = strip" \
+	"RANLIB = ranlib" \
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm -lpthread"
+
+# Linux x86 using Open MPI
+linux-ompi:
+	$(MAKE) all \
+	"ARCH = linux-ompi" \
+	"CC = mpicc" \
+	"CFLAGS = -m486 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI $(MISCFLAGS)" \
 	"AR = ar" \
 	"ARFLAGS = r" \
 	"STRIP = strip" \
 	"RANLIB = ranlib" \
-	"LIBS = -L. -L$(LAMHOME)/lib -ltachyon -lmpi $(MISCLIB) -lm -lpthread"
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm"
+
+# Linux AMD64/EM64T using Open MPI
+linux-ompi-64:
+	$(MAKE) all \
+	"ARCH = linux-ompi-64" \
+	"CC = mpicc" \
+	"CFLAGS = -m64 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI -DLP64 $(MISCFLAGS)" \
+	"AR = ar" \
+	"ARFLAGS = r" \
+	"STRIP = strip" \
+	"RANLIB = ranlib" \
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm"
+
+# Linux AMD64/EM64T using Open MPI, with OpenGL display
+linux-ompi-64-ogl:
+	$(MAKE) all \
+	"ARCH = linux-ompi-64-ogl" \
+	"CC = mpicc" \
+	"CFLAGS = -m64 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI -DLP64 -DUSEOPENGL $(LINUX_GLX_INCS) $(MISCFLAGS)" \
+	"AR = ar" \
+	"ARFLAGS = r" \
+	"STRIP = strip" \
+	"RANLIB = ranlib" \
+	"LIBS = -L. -ltachyon $(LINUX_GLX_LIBS) $(MISCLIB) -lm"
+
+# Linux x86 using Open MPI and threads
+linux-ompi-thr:
+	$(MAKE) all \
+	"ARCH = linux-ompi-thr" \
+	"CC = mpicc" \
+	"CFLAGS = -m486 -Wall -O6 -fomit-frame-pointer -ffast-math -DLinux -DMPI -DTHR -D_REENTRANT $(MISCFLAGS)" \
+	"AR = ar" \
+	"ARFLAGS = r" \
+	"STRIP = strip" \
+	"RANLIB = ranlib" \
+	"LIBS = -L. -ltachyon $(MISCLIB) -lm -lpthread"
 
 # Linux on iPaq, using gcc
 linux-ipaq:

