diff -ru hwloc-1.7-original/utils/hwloc-info.1in hwloc-1.7/utils/hwloc-info.1in
--- hwloc-1.7-original/utils/hwloc-info.1in	2013-03-26 21:00:18.000000000 +0100
+++ hwloc-1.7/utils/hwloc-info.1in	2013-05-09 09:19:06.397752380 +0200
@@ -148,7 +148,7 @@
 before reading this man page.  Most of the concepts described in
 hwloc(7) directly apply to the hwloc-calc utility.
 .
-..\" **************************
+.\" **************************
 .\"    Examples Section
 .\" **************************
 .SH EXAMPLES
