Index: orte/mca/plm/rsh/plm_rsh_component.c
===================================================================
--- orte/mca/plm/rsh/plm_rsh_component.c	(revision 29713)
+++ orte/mca/plm/rsh/plm_rsh_component.c	(working copy)
@@ -206,7 +206,7 @@
                              "%s plm:rsh: unable to be used: cannot find path "
                              "for launching agent \"%s\"\n", 
                              ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),
-                             orte_plm_globals.rsh_agent_argv[0]));
+                             (NULL == orte_plm_globals.rsh_agent_argv[0]) ? "NULL" : orte_plm_globals.rsh_agent_argv[0]));
         *module = NULL;
         return ORTE_ERROR;
     }

