Index: ompi/datatype/ompi_datatype_internal.h
===================================================================
--- ompi/datatype/ompi_datatype_internal.h	(revision 28543)
+++ ompi/datatype/ompi_datatype_internal.h	(working copy)
@@ -429,7 +429,7 @@
 #define OMPI_DATATYPE_INIT_DESC_PREDEFINED(TYPE, SIZE)                               \
     {                                                                                \
         1 /*length*/, 1 /*used*/,                                                    \
-        &(ompi_datatype_predefined_elem_desc[2 * OPAL_DATATYPE_ ## TYPE ## SIZE]) /*desc*/ \
+        &(opal_datatype_predefined_elem_desc[2 * OPAL_DATATYPE_ ## TYPE ## SIZE]) /*desc*/ \
     }
 
 /*
Index: ompi/datatype/ompi_datatype_module.c
===================================================================
--- ompi/datatype/ompi_datatype_module.c	(revision 28543)
+++ ompi/datatype/ompi_datatype_module.c	(working copy)
@@ -437,25 +437,9 @@
          * the OMPI predefined_elem_desc and update the fields there.
          */
         pDesc = &ompi_datatype_predefined_elem_desc[2 * i];
-        if( pDesc != datatype->super.desc.desc ) {
-            memcpy(pDesc, datatype->super.desc.desc, 2 * sizeof(dt_elem_desc_t));
-            datatype->super.desc.desc = pDesc;
-        } else {
-            datatype->super.desc.desc[0].elem.common.flags = OPAL_DATATYPE_FLAG_PREDEFINED |
-                                                             OPAL_DATATYPE_FLAG_DATA |
-                                                             OPAL_DATATYPE_FLAG_CONTIGUOUS |
-                                                             OPAL_DATATYPE_FLAG_NO_GAPS;
-            datatype->super.desc.desc[0].elem.common.type  = i;
-            datatype->super.desc.desc[0].elem.count        = 1;
-            datatype->super.desc.desc[0].elem.disp         = 0;
-            datatype->super.desc.desc[0].elem.extent       = datatype->super.size;
-
-            datatype->super.desc.desc[1].end_loop.common.flags    = 0;
-            datatype->super.desc.desc[1].end_loop.common.type     = OPAL_DATATYPE_END_LOOP;
-            datatype->super.desc.desc[1].end_loop.items           = 1;
-            datatype->super.desc.desc[1].end_loop.first_elem_disp = datatype->super.desc.desc[0].elem.disp;
-            datatype->super.desc.desc[1].end_loop.size            = datatype->super.size;
-        }
+        memcpy(pDesc, datatype->super.desc.desc, 2 * sizeof(dt_elem_desc_t));
+        datatype->super.desc.desc = pDesc;
+        datatype->super.opt_desc.desc = pDesc;
         /* Check if the data contain gaps */
         if( (datatype->super.ub - datatype->super.lb) != (OPAL_PTRDIFF_TYPE)datatype->super.size ) {
             datatype->super.desc.desc[0].elem.common.flags &= ~OPAL_DATATYPE_FLAG_NO_GAPS;
