Index: ompi/mca/btl/scif/btl_scif_module.c
===================================================================
--- ompi/mca/btl/scif/btl_scif_module.c	(revision 31667)
+++ ompi/mca/btl/scif/btl_scif_module.c	(working copy)
@@ -78,7 +78,13 @@
     rc = scif_get_nodeIDs (NULL, 0, &mca_btl_scif_module.port_id.node);
     if (-1 == rc) {
         BTL_VERBOSE(("btl/scif error getting node id of this node"));
+        scif_close (mca_btl_scif_module.scif_fd);
+        mca_btl_scif_module.scif_fd = -1;
         return OMPI_ERROR;
+    } else if (1 == rc) {
+        BTL_VERBOSE(("btl/scif no MIC detected"));
+        mca_btl_scif_module.scif_fd = -1;
+        return OMPI_ERROR;
     }
 
     /* Listen for connections */

