--- origsrc/hwloc-1.11.2/Makefile.am	2015-06-02 11:12:40.000000000 +0200
+++ Makefile.am	2016-04-12 21:54:05.390102200 +0200
@@ -21,11 +21,9 @@ DIST_SUBDIRS = $(SUBDIRS)
 
 # Only install the pkg file if we're building in standalone mode (and not on Windows)
 if HWLOC_BUILD_STANDALONE
-if !HWLOC_HAVE_WINDOWS
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = hwloc.pc
 endif
-endif
 
 # Only install the valgrind suppressions file if we're building in standalone mode
 if HWLOC_BUILD_STANDALONE

