--- origsrc/hwloc-1.11.3rc1/utils/lstopo/lstopo-draw.c	2016-04-13 09:10:40.000000000 +0200
+++ src/hwloc-1.11.3rc1/utils/lstopo/lstopo-draw.c	2016-04-16 23:29:54.402138800 +0200
@@ -1239,7 +1239,7 @@ fig(hwloc_topology_t topology, struct dr
       /* Display the hostname, but only if we're showing *this*
          system */
     if (hwloc_topology_is_thissystem(topology)) {
-#ifdef HWLOC_WIN_SYS
+#if defined(HWLOC_WIN_SYS) && !defined(__CYGWIN__)
       GetComputerName(hostname, &hostname_size);
 #else
       gethostname(hostname, hostname_size);

