diff --git a/ompi/mca/io/romio/romio/adio/ad_nfs/ad_nfs_read.c b/ompi/mca/io/romio/romio/adio/ad_nfs/ad_nfs_read.c
index 16f3a4d..2577f13 100644
--- a/ompi/mca/io/romio/romio/adio/ad_nfs/ad_nfs_read.c
+++ b/ompi/mca/io/romio/romio/adio/ad_nfs/ad_nfs_read.c
@@ -457,13 +457,14 @@ void ADIOI_NFS_ReadStrided(ADIO_File fd, void *buf, int count,
 	}
 	else {
 /* noncontiguous in memory as well as in file */
+            ADIO_Offset i;
 
 	    ADIOI_Flatten_datatype(datatype);
 	    flat_buf = ADIOI_Flatlist;
 	    while (flat_buf->type != datatype) flat_buf = flat_buf->next;
 
 	    k = num = buf_count = 0;
-	    i = (int) (flat_buf->indices[0]);
+	    i = flat_buf->indices[0];
 	    j = st_index;
 	    off = offset;
 	    n_filetypes = st_n_filetypes;
@@ -508,8 +509,8 @@ void ADIOI_NFS_ReadStrided(ADIO_File fd, void *buf, int count,
 
 		    k = (k + 1)%flat_buf->count;
 		    buf_count++;
-		    i = (int) (buftype_extent*(buf_count/flat_buf->count) +
-			flat_buf->indices[k]); 
+		    i = buftype_extent*(buf_count/flat_buf->count) +
+			flat_buf->indices[k]; 
 		    new_brd_size = flat_buf->blocklens[k];
 		    if (size != frd_size) {
 			off += size;

