Index: ompi/runtime/ompi_mpi_init.c
===================================================================
--- ompi/runtime/ompi_mpi_init.c	(revision 32669)
+++ ompi/runtime/ompi_mpi_init.c	(working copy)
@@ -17,6 +17,8 @@
  * Copyright (c) 2011      Sandia National Laboratories. All rights reserved.
  * Copyright (c) 2012-2013 Inria.  All rights reserved.
  * Copyright (c) 2014      Intel, Inc. All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  *
  * $COPYRIGHT$
  * 
@@ -58,7 +60,9 @@
 #include "opal/mca/pmix/pmix.h"
 
 #include "ompi/constants.h"
+#if OMPI_BUILD_FORTRAN_BINDINGS
 #include "ompi/mpi/fortran/base/constants.h"
+#endif
 #include "ompi/runtime/mpiruntime.h"
 #include "ompi/runtime/params.h"
 #include "ompi/communicator/communicator.h"
Index: ompi/mpi/c/Makefile.am
===================================================================
--- ompi/mpi/c/Makefile.am	(revision 32669)
+++ ompi/mpi/c/Makefile.am	(working copy)
@@ -15,6 +15,8 @@
 # Copyright (c) 2012-2013 Inria.  All rights reserved.
 # Copyright (c) 2013      Los Alamos National Security, LLC. All rights
 #                         reserved.
+# Copyright (c) 2014      Research Organization for Information Science
+#                         and Technology (RIST). All rights reserved.
 # $COPYRIGHT$
 # 
 # Additional copyrights may follow
@@ -103,7 +105,6 @@
         cart_sub.c \
         close_port.c \
         comm_accept.c \
-        comm_c2f.c \
         comm_call_errhandler.c \
         comm_compare.c \
         comm_connect.c \
@@ -116,7 +117,6 @@
         comm_dup.c \
         comm_dup_with_info.c \
         comm_idup.c \
-        comm_f2c.c \
         comm_free.c \
         comm_free_keyval.c \
         comm_get_attr.c \
@@ -145,9 +145,7 @@
         comm_test_inter.c \
 	compare_and_swap.c \
         dims_create.c \
-        errhandler_c2f.c \
         errhandler_create.c \
-        errhandler_f2c.c \
         errhandler_free.c \
         errhandler_get.c \
         errhandler_set.c \
@@ -179,11 +177,9 @@
         graphdims_get.c \
         grequest_complete.c \
         grequest_start.c \
-        group_c2f.c \
         group_compare.c \
         group_difference.c \
         group_excl.c \
-        group_f2c.c \
         group_free.c \
         group_incl.c \
         group_intersection.c \
@@ -196,11 +192,9 @@
         ibsend.c \
         improbe.c \
         imrecv.c \
-        info_c2f.c \
         info_create.c \
         info_delete.c \
         info_dup.c \
-        info_f2c.c \
         info_free.c \
         info_get.c \
         info_get_nkeys.c \
@@ -221,8 +215,6 @@
         keyval_create.c \
         keyval_free.c \
         lookup_name.c \
-        message_f2c.c \
-        message_c2f.c \
         mprobe.c \
         mrecv.c \
         neighbor_allgather.c \
@@ -235,10 +227,8 @@
         ineighbor_alltoallv.c \
         neighbor_alltoallw.c \
         ineighbor_alltoallw.c \
-        op_c2f.c \
         op_commutative.c \
         op_create.c \
-        op_f2c.c \
         op_free.c \
         open_port.c \
         pack_external.c \
@@ -259,8 +249,6 @@
         ireduce_scatter.c \
         reduce_scatter_block.c \
         ireduce_scatter_block.c \
-        request_c2f.c \
-        request_f2c.c \
         request_free.c \
         request_get_status.c \
 	rget.c \
@@ -282,8 +270,6 @@
         ssend.c \
         start.c \
         startall.c \
-        status_c2f.c \
-        status_f2c.c \
         status_set_cancelled.c \
         status_set_elements.c \
         status_set_elements_x.c \
@@ -293,7 +279,6 @@
         test_cancelled.c \
         testsome.c \
         topo_test.c \
-        type_c2f.c \
         type_commit.c \
         type_contiguous.c \
         type_create_darray.c \
@@ -311,7 +296,6 @@
         type_delete_attr.c \
         type_dup.c \
         type_extent.c \
-        type_f2c.c \
         type_free.c \
         type_free_keyval.c \
         type_get_attr.c \
@@ -349,7 +333,6 @@
 	win_allocate.c \
 	win_allocate_shared.c \
 	win_attach.c \
-        win_c2f.c \
         win_call_errhandler.c \
         win_complete.c  \
         win_create_errhandler.c \
@@ -358,7 +341,6 @@
 	win_create_dynamic.c \
         win_delete_attr.c \
 	win_detach.c \
-        win_f2c.c \
         win_fence.c \
 	win_flush.c \
 	win_flush_all.c \
@@ -386,14 +368,36 @@
 	win_unlock_all.c \
         win_wait.c
 
+if OMPI_BUILD_FORTRAN_BINDINGS
+libmpi_c_mpi_la_SOURCES += \
+        comm_c2f.c \
+        comm_f2c.c \
+        errhandler_c2f.c \
+        errhandler_f2c.c \
+        group_c2f.c \
+        group_f2c.c \
+        info_f2c.c \
+        info_c2f.c \
+        message_c2f.c \
+        message_f2c.c \
+        op_c2f.c \
+        op_f2c.c \
+        request_c2f.c \
+        request_f2c.c \
+        status_c2f.c \
+        status_f2c.c \
+        type_c2f.c \
+        type_f2c.c \
+        win_c2f.c \
+        win_f2c.c
+endif
+
 if OMPI_PROVIDE_MPI_FILE_INTERFACE
 libmpi_c_mpi_la_SOURCES += \
-        file_c2f.c \
         file_call_errhandler.c \
         file_close.c \
         file_create_errhandler.c \
         file_delete.c \
-        file_f2c.c \
         file_get_amode.c \
         file_get_atomicity.c \
         file_get_byte_offset.c \
@@ -446,8 +450,15 @@
         file_write_ordered_end.c \
         file_write_shared.c \
         register_datarep.c
+
+if OMPI_BUILD_FORTRAN_BINDINGS
+libmpi_c_mpi_la_SOURCES += \
+        file_c2f.c \
+        file_f2c.c
 endif
 
+endif
+
 # Conditionally install the header files
 
 if WANT_INSTALL_HEADERS
Index: ompi/mpi/c/profile/Makefile.am
===================================================================
--- ompi/mpi/c/profile/Makefile.am	(revision 32669)
+++ ompi/mpi/c/profile/Makefile.am	(working copy)
@@ -16,6 +16,8 @@
 # Copyright (c) 2012-2013 Inria.  All rights reserved.
 # Copyright (c) 2013      Los Alamos National Security, LLC. All rights
 #                         reserved.
+# Copyright (c) 2014      Research Organization for Information Science
+#                         and Technology (RIST). All rights reserved.
 # $COPYRIGHT$
 # 
 # Additional copyrights may follow
@@ -87,7 +89,6 @@
         pcart_sub.c \
         pclose_port.c \
         pcomm_accept.c \
-        pcomm_c2f.c \
         pcomm_call_errhandler.c \
         pcomm_compare.c \
         pcomm_connect.c \
@@ -100,7 +101,6 @@
         pcomm_dup.c \
         pcomm_dup_with_info.c \
         pcomm_idup.c \
-        pcomm_f2c.c \
         pcomm_free.c \
         pcomm_free_keyval.c \
         pcomm_get_attr.c \
@@ -129,9 +129,7 @@
         pcomm_test_inter.c \
 	pcompare_and_swap.c \
         pdims_create.c \
-        perrhandler_c2f.c \
         perrhandler_create.c \
-        perrhandler_f2c.c \
         perrhandler_free.c \
         perrhandler_get.c \
         perrhandler_set.c \
@@ -163,11 +161,9 @@
         pgraphdims_get.c \
         pgrequest_complete.c \
         pgrequest_start.c \
-        pgroup_c2f.c \
         pgroup_compare.c \
         pgroup_difference.c \
         pgroup_excl.c \
-        pgroup_f2c.c \
         pgroup_free.c \
         pgroup_incl.c \
         pgroup_intersection.c \
@@ -180,11 +176,9 @@
         pibsend.c \
 	pimprobe.c \
 	pimrecv.c \
-        pinfo_c2f.c \
         pinfo_create.c \
         pinfo_delete.c \
         pinfo_dup.c \
-        pinfo_f2c.c \
         pinfo_free.c \
         pinfo_get.c \
         pinfo_get_nkeys.c \
@@ -205,8 +199,6 @@
         pkeyval_create.c \
         pkeyval_free.c \
         plookup_name.c \
-        pmessage_f2c.c \
-        pmessage_c2f.c \
         pmprobe.c \
         pmrecv.c \
         pneighbor_allgather.c \
@@ -219,10 +211,8 @@
         pineighbor_alltoallv.c \
         pneighbor_alltoallw.c \
         pineighbor_alltoallw.c \
-        pop_c2f.c \
         pop_create.c \
         pop_commutative.c \
-        pop_f2c.c \
         pop_free.c \
         popen_port.c \
         ppack_external.c \
@@ -243,8 +233,6 @@
         pireduce_scatter.c \
         preduce_scatter_block.c \
         pireduce_scatter_block.c \
-        prequest_c2f.c \
-        prequest_f2c.c \
         prequest_free.c \
         prequest_get_status.c \
 	prget.c \
@@ -266,8 +254,6 @@
         pssend.c \
         pstart.c \
         pstartall.c \
-        pstatus_c2f.c \
-        pstatus_f2c.c \
         pstatus_set_cancelled.c \
         pstatus_set_elements.c \
         pstatus_set_elements_x.c \
@@ -277,7 +263,6 @@
         ptest_cancelled.c \
         ptestsome.c \
         ptopo_test.c \
-        ptype_c2f.c \
         ptype_commit.c \
         ptype_contiguous.c \
         ptype_create_darray.c \
@@ -295,7 +280,6 @@
         ptype_delete_attr.c \
         ptype_dup.c \
         ptype_extent.c \
-        ptype_f2c.c \
         ptype_free.c \
         ptype_free_keyval.c \
         ptype_get_attr.c \
@@ -333,7 +317,6 @@
 	pwin_allocate.c \
 	pwin_allocate_shared.c \
 	pwin_attach.c \
-        pwin_c2f.c \
         pwin_call_errhandler.c \
         pwin_complete.c  \
         pwin_create_errhandler.c \
@@ -342,7 +325,6 @@
 	pwin_create_dynamic.c \
         pwin_delete_attr.c \
 	pwin_detach.c \
-        pwin_f2c.c \
         pwin_fence.c \
 	pwin_flush.c \
 	pwin_flush_all.c \
@@ -370,14 +352,36 @@
 	pwin_unlock_all.c \
         pwin_wait.c
 
+if OMPI_BUILD_FORTRAN_BINDINGS
+nodist_libmpi_c_pmpi_la_SOURCES += \
+        pcomm_c2f.c \
+        pcomm_f2c.c \
+        perrhandler_c2f.c \
+        perrhandler_f2c.c \
+        pgroup_c2f.c \
+        pgroup_f2c.c \
+        pinfo_c2f.c \
+        pinfo_f2c.c \
+        pmessage_c2f.c \
+        pmessage_f2c.c \
+        pop_c2f.c \
+        pop_f2c.c \
+        prequest_c2f.c \
+        prequest_f2c.c \
+        pstatus_c2f.c \
+        pstatus_f2c.c \
+        ptype_c2f.c \
+        ptype_f2c.c \
+        pwin_c2f.c \
+        pwin_f2c.c
+endif
+
 if OMPI_PROVIDE_MPI_FILE_INTERFACE
 nodist_libmpi_c_pmpi_la_SOURCES += \
-        pfile_c2f.c \
         pfile_call_errhandler.c \
         pfile_close.c \
         pfile_create_errhandler.c \
         pfile_delete.c \
-        pfile_f2c.c \
         pfile_get_amode.c \
         pfile_get_atomicity.c \
         pfile_get_byte_offset.c \
@@ -430,8 +434,15 @@
         pfile_write_ordered_end.c \
         pfile_write_shared.c \
         pregister_datarep.c
+
+if OMPI_BUILD_FORTRAN_BINDINGS
+nodist_libmpi_c_pmpi_la_SOURCES += \
+        pfile_c2f.c \
+        pfile_f2c.c
 endif
 
+endif
+
 #
 # Sym link in the sources from the real MPI directory
 #
Index: ompi/request/grequest.c
===================================================================
--- ompi/request/grequest.c	(revision 32669)
+++ ompi/request/grequest.c	(working copy)
@@ -11,6 +11,8 @@
  *                         All rights reserved.
  * Copyright (c) 2006-2012 Cisco Systems, Inc.  All rights reserved.
  * Copyright (c) 2009      Sun Microsystems, Inc.  All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  * 
  * Additional copyrights may follow
@@ -216,11 +218,15 @@
         if (g->greq_funcs_are_c) {
             rc = g->greq_query.c_query(g->greq_state, status);
         } else {
+#if OMPI_BUILD_FORTRAN_BINDINGS
             MPI_Fint ierr;
             MPI_Fint fstatus[sizeof(MPI_Status) / sizeof(int)];
             g->greq_query.f_query((MPI_Aint*)g->greq_state, fstatus, &ierr);
             MPI_Status_f2c(fstatus, status);
             rc = OMPI_FINT_2_INT(ierr);
+#else
+            rc = OMPI_ERROR;
+#endif
         }
     }
 
Index: config/ompi_setup_mpi_fortran.m4
===================================================================
--- config/ompi_setup_mpi_fortran.m4	(revision 32669)
+++ config/ompi_setup_mpi_fortran.m4	(working copy)
@@ -15,6 +15,8 @@
 # Copyright (c) 2006-2007 Los Alamos National Security, LLC.  All rights
 #                         reserved. 
 # Copyright (c) 2009      Oak Ridge National Labs.  All rights reserved.
+# Copyright (c) 2014      Research Organization for Information Science
+#                         and Technology (RIST). All rights reserved.
 # $COPYRIGHT$
 # 
 # Additional copyrights may follow
@@ -691,4 +693,10 @@
     # safe, too.
     AM_CONDITIONAL(OMPI_BUILD_FORTRAN_USEMPIF08_BINDINGS, 
                    [test $OMPI_BUILD_FORTRAN_USEMPIF08_BINDINGS -eq 1])
+
+    AM_CONDITIONAL(OMPI_BUILD_FORTRAN_BINDINGS,
+                   [test $OMPI_BUILD_FORTRAN_MPIFH_BINDINGS -eq 1 || \
+                    test $OMPI_BUILD_FORTRAN_USEMPI_BINDINGS -eq 1 || \
+                    test $OMPI_BUILD_FORTRAN_USEMPIF08_BINDINGS -eq 1])
+
 ])

