diff --git a/ompi/datatype/ompi_datatype_create_subarray.c b/ompi/datatype/ompi_datatype_create_subarray.c
index 016a119..e6d16be 100644
--- a/ompi/datatype/ompi_datatype_create_subarray.c
+++ b/ompi/datatype/ompi_datatype_create_subarray.c
@@ -36,7 +36,7 @@ int32_t ompi_datatype_create_subarray(int ndims,
                                       const ompi_datatype_t* oldtype,
                                       ompi_datatype_t** newtype)
 {
-    MPI_Datatype last_type;
+    ompi_datatype_t *last_type;
     int32_t i, step, end_loop;
     MPI_Aint size, displ, extent;
 
@@ -102,6 +102,9 @@ int32_t ompi_datatype_create_subarray(int ndims,
     *newtype = ompi_datatype_create( last_type->super.desc.used );
     ompi_datatype_add( *newtype, last_type, 1, displ * extent, size * extent);
     ompi_datatype_destroy( &last_type );
+    last_type = *newtype;
+    ompi_datatype_create_resized(last_type, 0, size * extent, newtype);
+    ompi_datatype_destroy( &last_type );
 
     return OMPI_SUCCESS;
 }

