--- openmpi-1.4.3.original/configure	2010-10-05 15:48:18.000000000 -0700
+++ openmpi-1.4.3.patched/configure	2011-11-04 15:01:37.000000000 -0700
@@ -10504,7 +10504,8 @@
 
     ompi_ident="string_not_coincidentally_inserted_by_the_compiler"
     cat > conftest.c <<EOF
-#pragma ident "$ompi_ident"
+#define IDENT "$ompi_ident"
+#pragma ident $IDENT
 int main(int argc, char** argv);
 int main(int argc, char** argv) { return 0; }
 EOF
@@ -26217,7 +26218,8 @@
 
     ompi_ident="string_not_coincidentally_inserted_by_the_compiler"
     cat > conftest.cc <<EOF
-#pragma ident "$ompi_ident"
+#define IDENT "$ompi_ident"
+#pragma ident $IDENT
 int main(int argc, char** argv);
 int main(int argc, char** argv) { return 0; }
 EOF
@@ -34690,6 +34692,11 @@
     { $as_echo "$as_me:$LINENO: checking if $CXX supports GCC inline assembly" >&5
 $as_echo_n "checking if $CXX supports GCC inline assembly... " >&6; }
 
+    if test "$ompi_cv_cxx_compiler_vendor" = "portland group" ; then
+        # PGI seems to have some issues with our inline assembly.
+        # Disable for now.
+        asm_result="no (Portland Group)"
+    else
     case $host in
         *-aix*)
             # the AIX compilers and linkers really don't do gcc
@@ -34813,6 +34820,7 @@
 rm -f core conftest.err conftest.$ac_objext conftest_ipa8_conftest.oo \
       conftest$ac_exeext conftest.$ac_ext
     fi
+    fi
 
     { $as_echo "$as_me:$LINENO: result: $asm_result" >&5
 $as_echo "$asm_result" >&6; }
@@ -47269,7 +47277,7 @@
     # there are some places in the code where we have to have *something*.
 
 cat >>confdefs.h <<_ACEOF
-#define OMPI_HAVE_FORTRAN_REAL16 $ofc_have_type
+#define OMPI_HAVE_FORTRAN_REAL16 ( $ofc_have_type && OMPI_REAL16_MATCHES_C )
 _ACEOF
 
 
@@ -50744,7 +50752,7 @@
     # there are some places in the code where we have to have *something*.
 
 cat >>confdefs.h <<_ACEOF
-#define OMPI_HAVE_FORTRAN_COMPLEX32 $ofc_have_type
+#define OMPI_HAVE_FORTRAN_COMPLEX32 ( $ofc_have_type && OMPI_REAL16_MATCHES_C )
 _ACEOF
 
 
@@ -58335,7 +58343,7 @@
 
 
 cat >>confdefs.h <<_ACEOF
-#define OMPI_HAVE_F90_REAL16 $ofc_have_type
+#define OMPI_HAVE_F90_REAL16 ( $ofc_have_type && OMPI_REAL16_MATCHES_C )
 _ACEOF
 
 
@@ -60152,7 +60160,7 @@
 
 
 cat >>confdefs.h <<_ACEOF
-#define OMPI_HAVE_F90_COMPLEX32 $ofc_have_type
+#define OMPI_HAVE_F90_COMPLEX32 ( $ofc_have_type && OMPI_REAL16_MATCHES_C )
 _ACEOF
 
 

