Index: configure.ac
===================================================================
--- configure.ac	(revision 26462)
+++ configure.ac	(working copy)
@@ -48,14 +48,18 @@
 AC_CONFIG_AUX_DIR(config)
 AC_CONFIG_MACRO_DIR(config)
 
+# Get our platform support file.  This has to be done very, very early
+# because it twiddles random bits of autoconf
+OMPI_LOAD_PLATFORM
+
 #
+#
 # Start it up
 #
 
 OMPI_CONFIGURE_SETUP
 ompi_show_title "Configuring project_name_long"
 
-#
 # Setup some things that must be done before AM-INIT-AUTOMAKE
 #
 
@@ -69,9 +73,6 @@
        AC_MSG_WARN([Proceed at your own risk!])])
 AC_USE_SYSTEM_EXTENSIONS
 
-# Get our platform support file.  This has to be done very, very early
-# because it twiddles random bits of autoconf
-OMPI_LOAD_PLATFORM
 
 #
 # Init automake

