diff --git a/config/hwloc.m4 b/config/hwloc.m4
index 2fdc198..037c9b4 100644
--- a/config/hwloc.m4
+++ b/config/hwloc.m4
@@ -640,7 +640,7 @@ EOF])
 
     # PCI support
     hwloc_pci_happy=no
-    if test "x$enable_pci" != xno; then
+    if test "x$enable_pci" != xno -a "x$enable_libpci" != "xyes"; then
       hwloc_pci_happy=yes
       HWLOC_PKG_CHECK_MODULES([PCIACCESS], [pciaccess], [pci_system_init], [:], [hwloc_pci_happy=no])
       if test x$hwloc_pci_happy = xyes; then hwloc_pci_lib=pciaccess; fi
@@ -686,11 +686,18 @@ EOF])
                     [hwloc_pci_happy=no])])
             ], [hwloc_pci_happy=no])
         ])
-        if test x$hwloc_pci_happy = xyes; then hwloc_pci_lib=pciutils; fi
+        if test x$hwloc_pci_happy = xyes; then
+	  if test x$enable_libpci = xyes; then
+	    hwloc_pci_lib=pciutils
+	  else
+	    hwloc_pci_happy=no
+	    hwloc_warn_may_use_libpci=yes
+	  fi
+	fi
     fi
     AC_SUBST(HWLOC_PCI_LIBS)
     # If we asked for pci support but couldn't deliver, fail
-    AS_IF([test "$enable_pci" = "yes" -a "$hwloc_pci_happy" = "no"],
+    AS_IF([test "$enable_pci" = "yes" -a "$hwloc_pci_happy" = "no" -a "$hwloc_warn_may_use_libpci" != "yes"],
           [AC_MSG_WARN([Specified --enable-pci switch, but could not])
            AC_MSG_WARN([find appropriate support])
            AC_MSG_ERROR([Cannot continue])])
diff --git a/config/hwloc_internal.m4 b/config/hwloc_internal.m4
index 3a79ae6..5970185 100644
--- a/config/hwloc_internal.m4
+++ b/config/hwloc_internal.m4
@@ -64,6 +64,9 @@ AC_DEFUN([HWLOC_DEFINE_ARGS],[
     AC_ARG_ENABLE([pci],
                   AS_HELP_STRING([--disable-pci],
                                  [Disable the PCI device discovery using libpci]))
+    AC_ARG_ENABLE([libpci],
+		  AS_HELP_STRING([--enable-libpci],
+				 [Use libpci for PCI support. Note that this may taint hwloc as GPL.]))
 
     # OpenCL?
     AC_ARG_ENABLE([opencl],
diff --git a/configure.ac b/configure.ac
index d85de3b..c9f5064 100644
--- a/configure.ac
+++ b/configure.ac
@@ -206,6 +206,16 @@ append_env() {
     eval "[$]1=\"$new_value\""
 }
 
+# Warn about PCI support
+if test x$hwloc_warn_may_use_libpci = xyes; then
+  echo
+  echo "**********************************************************************"
+  echo "PCI support could not be enable because libpciaccess is not available."
+  echo "libpci could be used instead but it may taint hwloc as GPL."
+  echo "Add --enable-libpci to enable it."
+  echo "**********************************************************************"
+fi
+
 # Show which optional support we'll be building
 hwloc_xml_status=basic
 AS_IF([test "$hwloc_libxml2_happy" = "yes"], [hwloc_xml_status=full])

