--- /home/hpcmtt/mtt-trunk/client/mtt-relay.orig        2013-09-30 11:35:59.637400212 +0200
+++ /home/hpcmtt/mtt-trunk/client/mtt-relay     2013-09-30 11:45:19.496180413 +0200
@@ -93,6 +93,10 @@
     #     exit;
     # }
 
+    my $pidfilename = "/var/run/mtt-relay.pid";
+    if (-e $pidfilename) {
+        exit 1;
+    }
     my $master = new HTTP::Daemon
             LocalAddr => $HOST, LocalPort => $PORT or 
             die "Problem creating an HTTP::Daemon: $!";
@@ -103,6 +107,9 @@
         my $pid = fork();
         if($pid) {
             print "# Daemon Parent exiting\n";
+            open PIDFILE, ">$pidfilename";
+            print PIDFILE $pid;
+            close PIDFILE;
             exit 0;
         } else {
             print "# Daemon Child process continuing\n";
