Index: MTT/MPI/Install.pm
===================================================================
--- MTT/MPI/Install.pm	(revision 1123)
+++ MTT/MPI/Install.pm	(working copy)
@@ -248,6 +252,8 @@
                                 $MTT::Globals::Internals->{mpi_get_name} =
                                     $mpi_get_key;
                                 $MTT::Globals::Internals->{mpi_install_name} = $simple_section;
+				#OLEG#
+				MTT::DoCommand::Chdir($install_base);
                                 my $mpi_dir = _make_random_dir(4);
                                 MTT::DoCommand::Chdir($mpi_dir);
                             
