--- a/ompi/mca/btl/sm/btl_sm_component.c.orig	2013-02-22 11:50:16.000000000 -0500
+++ a/ompi/mca/btl/sm/btl_sm_component.c	2013-02-22 11:50:53.000000000 -0500
@@ -671,7 +671,7 @@
  */
 static int
 backing_store_init(mca_btl_sm_component_t *comp_ptr,
-                   orte_local_rank_t local_rank)
+                   ompi_local_rank_t local_rank)
 {
     int rc = OMPI_SUCCESS;
 
@@ -706,7 +706,7 @@
 {
     int num_local_procs = 0;
     mca_btl_base_module_t **btls = NULL;
-    orte_local_rank_t my_local_rank = OMPI_LOCAL_RANK_INVALID;
+    ompi_local_rank_t my_local_rank = OMPI_LOCAL_RANK_INVALID;
 #if OMPI_BTL_SM_HAVE_KNEM
     int rc;
 #endif /* OMPI_BTL_SM_HAVE_KNEM */
--- a/ompi/mca/btl/sm/btl_sm_component.c.orig	2013-02-21 13:28:07.000000000 -0500
+++ a/ompi/mca/btl/sm/btl_sm_component.c	2013-02-22 12:00:21.000000000 -0500
@@ -671,7 +671,7 @@
  */
 static int
 backing_store_init(mca_btl_sm_component_t *comp_ptr,
-                   orte_local_rank_t local_rank)
+                   ompi_local_rank_t local_rank)
 {
     int rc = OMPI_SUCCESS;
 
@@ -706,7 +706,7 @@
 {
     int num_local_procs = 0;
     mca_btl_base_module_t **btls = NULL;
-    orte_local_rank_t my_local_rank = OMPI_LOCAL_RANK_INVALID;
+    ompi_local_rank_t my_local_rank = OMPI_LOCAL_RANK_INVALID;
 #if OMPI_BTL_SM_HAVE_KNEM
     int rc;
 #endif /* OMPI_BTL_SM_HAVE_KNEM */
