Index: orte/mca/plm/base/plm_base_launch_support.c
===================================================================
--- orte/mca/plm/base/plm_base_launch_support.c	(revision 25056)
+++ orte/mca/plm/base/plm_base_launch_support.c	(working copy)
@@ -754,10 +754,7 @@
     ORTE_PROGRESSED_WAIT(app_launch_failed, jdata->num_launched, jdata->num_procs);
 
     /* cancel the lingering recv */
-    if (ORTE_SUCCESS != (rc = orte_rml.recv_cancel(ORTE_NAME_WILDCARD, ORTE_RML_TAG_APP_LAUNCH_CALLBACK))) {
-        ORTE_ERROR_LOG(rc);
-        return rc;
-    }
+    orte_rml.recv_cancel(ORTE_NAME_WILDCARD, ORTE_RML_TAG_APP_LAUNCH_CALLBACK);
     
     OPAL_OUTPUT_VERBOSE((5, orte_plm_globals.output,
                          "%s plm:base:report_launched all apps reported",
@@ -983,6 +980,12 @@
         }
     }
 
+    /* add the gid */
+    opal_argv_append(argc, argv, "--gid");
+    asprintf(&param, "%lu", (unsigned long)getgid());
+    opal_argv_append(argc, argv, param);
+    free(param);
+
     return ORTE_SUCCESS;
 }
 
Index: orte/orted/orted_main.c
===================================================================
--- orte/orted/orted_main.c	(revision 25056)
+++ orte/orted/orted_main.c	(working copy)
@@ -114,6 +114,7 @@
     int fail_delay;
     bool abort;
     int heartbeat;
+    char *gid;
 } orted_globals;
 
 /*
@@ -189,6 +190,10 @@
         NULL, OPAL_CMD_LINE_TYPE_STRING,
     "Create a new xterm window and display output from the specified ranks there" },
     
+    { NULL, NULL, NULL, '\0', "gid", "gid", 1,
+      &orted_globals.gid, OPAL_CMD_LINE_TYPE_STRING,
+    "Group ID to use" },
+    
     /* End of list */
     { NULL, NULL, NULL, '\0', NULL, NULL, 0,
       NULL, OPAL_CMD_LINE_TYPE_NULL, NULL }
@@ -206,6 +211,7 @@
     opal_buffer_t *buffer;
     char hostname[100];
     char *tmp_env_var = NULL;
+    gid_t gid;
 
     /* initialize the globals */
     memset(&orted_globals, 0, sizeof(orted_globals));
@@ -234,6 +240,15 @@
      */
     mca_base_cmd_line_process_args(cmd_line, &environ, &environ);
     
+    /* set the gid, if given */
+    if (NULL != orted_globals.gid) {
+        gid = (gid_t)strtoul(orted_globals.gid, NULL, 10);
+        if (0 > setgid(gid)) {
+            fprintf(stderr, "setgid FAILED\n");
+            return 1;
+        }
+    }
+
     /* Ensure that enough of OPAL is setup for us to be able to run */
     /*
      * NOTE: (JJH)

